---
title: "Gigascience_submission"
author: "Ralf"
date: "4/29/2021"
output: html_document
---

```{r setup, echo = T, results='hide'}

library(reshape2)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(Matrix)
library(jsonlite)
library(egg)
library(ComplexHeatmap)
library(svglite)
library(SCINA)
library(ggalluvial)
library(dplyr)

#### Define 
github_path="/media/Helios_scStorage/Ralf/Benchmark/Gigascience/git/BenchmarkAlignment/"
main_outpath="/media/Storage/Benchmark_test3/"
setwd(main_outpath)

#### State which datasets should be used ####
datasets <- c("pbmc_10x")#"Endothelial", "Cardiac", "HF",
datasets_nice <- c("PBMC")# "Endothelial", "Cardiac", "HF",
#### Which tool ####
tools <- c("cellranger6", "starsolo", "alevin-fry", "alevin", "kallisto")
tools_nice <- c("Cell Ranger 6", "STARsolo", "Alevin-fry", "Alevin", "Kallisto")
#### and which filtering ####
filtered <- c("results_filtered_anno")#, "results_unfiltered_anno"
```


```{r import function}

#### Create function to import expression matrices ####
read_data <- function(dataset_m, main_outpath, tool_m, filtered_m=NULL, sample_m, m_gene_col=2){
  path_in <- paste(main_outpath, "MappedReads", dataset_m, tool_m, filtered_m, sample_m, sep = "/")
  
  if (grepl("cellranger", tool_m)){
    print(paste0(path_in, "/outs/emptyDrops_feature_bc_matrix/"))
    data <- Read10X(paste0(path_in, "/outs/emptyDrops_feature_bc_matrix/"), 
                    gene.column = m_gene_col,
                    strip.suffix = T)
    
  }else if (tool_m=="starsolo"){
    print(paste0(path_in, "/Solo.out/Gene/filtered_matrix_emptyDrops/"))
    data <- Read10X(paste0(path_in, "/Solo.out/Gene/filtered_matrix_emptyDrops/"), gene.column = m_gene_col)
    
  }else if (tool_m == "alevin-fry") {
    print(paste0(path_in, "/",sample_m, "_quant_res/filtered_matrix_emptyDrops/"))
    data <- Read10X(paste0(path_in, "/",sample_m, "_quant_res/filtered_matrix_emptyDrops/"), gene.column = m_gene_col)
    
  }else if (tool_m=="alevin"){
    if (m_gene_col==2){
      data <- import_alevin_mat(paste0(main_outpath, "/MappedReads/", dataset_m, "/", tool_m, "/", filtered_m, "/"), sample_m, "alevin/quants_mat_cols.txt")
    }else if (m_gene_col==1){
      data <- import_alevin_mat(paste0(main_outpath, "/MappedReads/", dataset_m, "/", tool_m, "/", filtered_m, "/"), sample_m, "alevin/quants_mat_cols_id.txt")
    }
    
  }else if (tool_m=="kallisto"){
    print(paste0(path_in, "/counting/filtered_matrix_emptyDrops/"))
    data <- Read10X(paste0(path_in, "/counting/filtered_matrix_emptyDrops/"), gene.column = m_gene_col)
  }
}

import_alevin_mat <- function(sample_path, sample_name, m_genes_file){
  dir <- paste0(sample_path,sample_name,"/")
  print(dir)
  barcode.file <- file.path(dir, "alevin/quants_mat_rows.txt")
  gene.file <- file.path(dir, m_genes_file)
  matrix.file <- file.path(dir, "alevin/quants_mat.gz")
  #var.file <- file.path(dir, "alevin/quants_var_mat.gz")

  cell.names <- readLines(barcode.file)
  gene.names <- readLines(gene.file)

  readAlevinFast <- function(matrix.file, gene.names, cell.names) {
    num.cells <- length(cell.names)
    num.genes <- length(gene.names)
    mat <- fishpond::readEDS(num.genes, num.cells, matrix.file)
    dimnames(mat) <- list(gene.names, cell.names)
    return(mat)
  }
  readAlevinFast(matrix.file, gene.names, cell.names)
}

sem <- function(x){
  sd(x)/sqrt(length(x))
}
```

```{r save seurat objects, echo= T, results='hide'}

save_seurat_objects <- function(dataset, main_outpath, tools, filtered, set_gene_col=2){
  filtered_data <- list()
  seurat_list <- list()
  for (i in datasets){
  
    if (i == "Cardiac"){
      print(i)
      sample_names <- paste0(c("MF1701"), c(0, 3:8))
    } else if(i == "Endothelial"){
      sample_names <- c("Brain", "Colon", "Heart", "Kidney", "Liver", "Lung", "muscle_EDL", "muscle_Soleus", "Small_Intestine", "Spleen", "Testis")
    } else if(i == "pbmc_10x"){
      sample_names <- paste0(c("sample"), c(1:4))
    }else if(i == "HF"){
      sample_names <- paste0("103837-001-00", 1:5)
    }
  
    for (j in tools){
    
      print(j)
      #if(j == "cellranger5" && i == "Endothelial") next
    
      for (k in filtered){
      
        for (l in sample_names){
          if (length(filtered)==2){
            id_name <- paste(l,j,i,k,sep="__")
          }else{
            id_name <- paste(l,j,i,sep="__")
          }
          
          filtered_data[[id_name]] <- read_data(i,main_outpath, j,k,l,m_gene_col = set_gene_col)
          seurat_list[[id_name]] <- CreateSeuratObject(filtered_data[[id_name]], project = id_name)
          seurat_list[[id_name]][["percent.mt"]] <- PercentageFeatureSet(seurat_list[[id_name]], 
                                                                         pattern = "^MT-|^mt-")
        }
      }
    }
  }
  return(seurat_list)
}



pbmc_cardiac_endo <- save_seurat_objects(datasets, main_outpath, tools, filtered)
names(pbmc_cardiac_endo) <- gsub("pbmc_10x","PBMC", names(pbmc_cardiac_endo))
```


```{r Figure 1}
plot_mapping_performance <- function(plot_data_df, max_limit, y_name){
  ggplot(data = plot_data_df, aes(Dataset, y = mean, fill = Mapper)) +
    geom_col(position = "dodge2", colour="grey30", show.legend = T, width = 0.7) +
    geom_errorbar(aes(ymin=mean-sem,ymax=mean+sem), width = 0.4, position = position_dodge(.7)) +
    scale_fill_manual(values = c("#8dd3c7", "#bebada", "#fb8072", "#fdb462", "#80b1d3")) +
    #rev(RColorBrewer::brewer.pal(4,'Blues'))
    #scale_fill_hue(direction = -1, h.start = 90) +
    scale_y_continuous(expand = c(0, 0),limits = c(0,max_limit)) +#+,limits = c(0,100),breaks = seq(0,100,20)
    xlab("") + ylab(y_name) +
    theme(axis.text.x = element_text(angle = 60,vjust=+1.05 ,hjust = +1.1),
          text = element_text(size = 14),
          panel.background = element_rect(fill = "white", colour = "grey70"),
          panel.grid = element_line(colour = "grey90"),
          panel.grid.major.x = element_blank(),
          plot.margin = unit(x=c(10,5,0,1),units = "mm"))
}

Figure_1 <- list()

#### runtime #####
get_runtimes <- function(dataset_m){
  print(dataset_m)
  runtimes <- lapply(tools, function(x) {
    print(x)
    main_path <- paste("MappedReads/", dataset_m, x, "results_filtered_anno/", sep = "/")
    runtime_paths <- dir(main_path, pattern = "runtime", ignore.case = T, recursive = T)
    #print(runtime_paths)
    if (x == "alevin-fry" | dataset_m == "HF" && x=="kallisto"){
      runtime_paths <- runtime_paths[grep("runtime_sec|runtime_empty_drops.txt", runtime_paths)]
    }else if (x %in% c("cellranger6", "starsolo")){
      runtime_paths <- runtime_paths[grep("runtime.txt|runtime_empty_drops.txt", runtime_paths)]

      
    }
    if (x != "alevin"){
      test_split <- strsplit(runtime_paths[seq(1,length(runtime_paths), by=2)],"/")
      sample_names <- lapply(test_split, function(x) x[1])
    }else{
      test_split <- strsplit(runtime_paths, "/")
      sample_names <- lapply(test_split, function(x) x[1])
    }

    runtime_list <- list()
    for (i in sample_names){
        test_grep<- runtime_paths[grep(i, runtime_paths)]
        run <- lapply(test_grep, function(p) {
      
        r <- readLines(paste0(main_path, p))
        if (grepl("sec",p)  && (x == "alevin-fry" | x=="kallisto" )){
            print("second")
            print(r)
            r <- as.integer(r)/60
        
        }else{
            #grep line with real runtime
            r <- unlist(strsplit(r[grep("real", r)], "\t|m|[.]|,"))
            r <- as.integer(r[2]) + as.integer(r[3])/60
        }
      
        r
         })
    runtime_list[[paste(i)]] <- sum(unlist(run))
    }
    runtime_list
  })
  names(runtimes) <- tools
  runtimes
}

all_runtimes <- lapply(datasets, function(x) {
  get_runtimes(x)
})
names(all_runtimes) <- datasets


# parse data for ggplot
all_runtimes_melt <- melt(all_runtimes)
colnames(all_runtimes_melt) <- c("Runtime", "sample_no", "Mapper", "Dataset")

runtime_mean <- aggregate(Runtime ~ Mapper + Dataset, data = all_runtimes_melt, mean)
runtime_sd <- aggregate(Runtime ~ Mapper + Dataset, data = all_runtimes_melt, sem)

runtime_mean$sem <- runtime_sd$Runtime
runtime_mean[3:4] <- runtime_mean[3:4]/60
colnames(runtime_mean)[3] <- "mean"
runtime_mean$Mapper <- factor(runtime_mean$Mapper, levels = tools)
levels(runtime_mean$Mapper) <- tools_nice
runtime_mean$Dataset <- factor(runtime_mean$Dataset)
levels(runtime_mean$Dataset) <- datasets_nice
Figure_1[["runtime"]] <- plot_mapping_performance(runtime_mean, 4, "Runtime (h)")




# create dataset for plotting
create_plot_data <- function(data, datasets_nice, tools, tools_nice){
  data_melt <- melt(data)
  meta_data <- do.call(rbind, strsplit(data_melt$L1, "__"))
  #meta_data[,2] <- do.call(rbind, strsplit(meta_data[,2], "results_filtered_anno/"))[,2]
  data_melt <- cbind(data_melt, meta_data)
  colnames(data_melt) <- c("value", "L1", "Sample", "Mapper", "Dataset")
  data_mean <- aggregate(value ~ Dataset + Mapper, data = data_melt, mean)
  data_sem <- aggregate(value ~ Dataset + Mapper, data = data_melt, sem)
  data_mean$sem <- data_sem$value
  #print((data_mean))
  # make data pretty
  colnames(data_mean)[3] <- "mean"
  #levels(data_mean$Dataset)[3] <- "PBMC"
  print(str(data_mean))
  data_mean$Dataset <- factor(data_mean$Dataset, levels = datasets_nice)
  data_mean$Mapper <- factor(data_mean$Mapper, levels = tools)
  levels(data_mean$Mapper) <- tools_nice
  
  return(data_mean)
}

#### Genes per cell ####
genes_per_cell <- lapply(pbmc_cardiac_endo, function(x) x$nFeature_RNA)

plot_data_genes_per_cell <- create_plot_data(genes_per_cell, datasets_nice, tools, tools_nice)

Figure_1[["genes_per_cell"]] <- plot_mapping_performance(plot_data_genes_per_cell, 2000, "Genes per cell")


#### Cell count ####
ncells <- lapply(pbmc_cardiac_endo, function(x) ncol(x))

plot_data_ncells <- create_plot_data(ncells, datasets_nice, tools, tools_nice)

Figure_1[["cell_count"]] <- plot_mapping_performance(plot_data_ncells, 15000, "Cell count")



#### Mapping rate ####

get_mapping_rate <- function(dataset_m){
mapping_rate <- lapply(tools, function(x) {
    print(x)
    main_path <- paste("MappedReads", dataset_m, x, "results_filtered_anno/", sep = "/")
    
    if (x == "alevin-fry"){
      runtime_paths <- dir(main_path, pattern = "*salmon_quant.log", ignore.case = T, recursive = T)
      lapply(runtime_paths, function(p) {
        m <- readLines(paste0(main_path, p))
        m <- strsplit(m[grep("*Selectively-aligned*", m)], "Selectively-aligned | total fragments out of ")
        m <- as.integer(unlist(m)[2:3])
        m <- (m[1] /m[2]) *100
      })
    }else if (x == "cellranger6"){
      runtime_paths <- dir(main_path, pattern = "*metrics_summary.csv$", ignore.case = T, recursive = T)
      lapply(runtime_paths, function(p) {
        m <- read.csv(paste(main_path, p, sep = "/"))
        as.numeric(substr(m$Reads.Mapped.Confidently.to.Transcriptome, 1, 4))
      })
    }else if (x == "starsolo"){
      runtime_paths <- dir(main_path, pattern = "*Summary.csv$", ignore.case = T, recursive = T)
      lapply(runtime_paths, function(p) {
        m <- read.csv(paste(main_path, p, sep = "/"), header = F)
        m[m$V1=="Reads Mapped to Transcriptome: Unique Genes",2] *100
      })
    }else if (x == "alevin"){
      runtime_paths <- dir(main_path, pattern = "*salmon_quant.log$", ignore.case = T, recursive = T)
        lapply(runtime_paths, function(p) {
        m <- readLines(paste0(main_path, p))
        m <- strsplit(m[grep("*Mapping rate*", m)], "= |%")
        m <- as.numeric(unlist(m)[2])
      })
    }else if (x == "kallisto"){
      runtime_paths <- dir(main_path, pattern = "*run_info.json$", ignore.case = T, recursive = T)
      lapply(runtime_paths, function(p) {
        m <- jsonlite::read_json(paste0(main_path, p))
        m$p_pseudoaligned
      })
    }
    
  })
names(mapping_rate) <- tools
mapping_rate
}


all_mapping_rates <- lapply(datasets, function(x) {
  get_mapping_rate(x)
})
names(all_mapping_rates) <- datasets


all_mapping_rates_melt <- melt(all_mapping_rates)
colnames(all_mapping_rates_melt) <- c("mapping_rate", "sample_no", "Mapper", "Dataset")
map_data <- aggregate(mapping_rate ~ Mapper + Dataset, data = all_mapping_rates_melt, mean)
map_data2 <- aggregate(mapping_rate ~ Mapper + Dataset, data = all_mapping_rates_melt, sem)
map_data$sem <- map_data2$mapping_rate
colnames(map_data)[3] <- "mean"
map_data$Dataset <- factor(map_data$Dataset, levels = datasets_nice)
map_data$Mapper <- factor(map_data$Mapper, levels = tools)
levels(map_data$Mapper) <- tools_nice

Figure_1[["mapping_rate"]] <- plot_mapping_performance(map_data, 100, "Mapping rate (%)")


#### Make figure 1 ####
ggpubr::ggarrange(plotlist = Figure_1, ncol = 2, nrow = 2, common.legend = T, 
                  legend = "bottom", labels = "AUTO", align = "hv")

ggsave(paste0(main_outpath, "plots_revision/Figure_1_hf.svg"), device = "svg", width = 8, height = 8)


```

```{r Suppl Figure 1}
#library(egg)

plot_data <- lapply(pbmc_cardiac_endo, function(x) {
  list(genes = x$nFeature_RNA,
       umi = x$nCount_RNA)
})

plot_data_long <- melt(plot_data)
plot_data_long <- cbind(plot_data_long, do.call(rbind,strsplit(plot_data_long$L1,"__")))
colnames(plot_data_long) <- c("value", "Metadata", "L1", "Samples", "Mapper", "Datasets")
plot_data_long$Metadata <-  factor(plot_data_long$Metadata, levels = c("umi", "genes"))
levels(plot_data_long$Metadata) <- c("UMI count per cell (log10)", "Genes per cell (log10)")
plot_data_long$Mapper <- factor(plot_data_long$Mapper, levels = tools)
levels(plot_data_long$Mapper) <- tools_nice
plot_data_long$Datasets <- factor(plot_data_long$Datasets, levels = datasets_nice)

p_density <- ggplot(plot_data_long, aes(x = value, colour=Mapper)) +
  geom_density(aes(linetype = Mapper),size = 1) +
  facet_grid(Datasets ~ Metadata) +
  scale_x_log10() +
  theme_article() +
  theme(text = element_text(size = 14), legend.position = "bottom") +
  xlab("") + ylab("Density")

p_density
ggsave(paste0(main_outpath, "plots_revision/Suppl_Figure_1_hf.svg"), p_density, device = "svg", height = 9, width = 7.5)

```


```{r Figure 2 and Suppl figure 2 Upset plot}
#library(ComplexHeatmap)
#### Set up Functions ####
get_counts <- function(dataset){
lapply(dataset, function(x) {
  cellcount <- Matrix::colSums(x)
  genecount <- Matrix::rowSums(x)
  cellcount <- cellcount[cellcount>0]
  genecount <- genecount[genecount>0]
  list(cells=names(cellcount), genes=names(genecount))
})
}

get_upset <- function(samples_m, data_m, type_m, tools_nice_m=tools_nice, 
                      tool_names=tools, sorting=T, dataset_name, get_only_mat=F){
  #tool_names <- c("cellranger6", "starsolo", "alevin", "alevin-fry", "kallisto")

  lapply(samples_m, function(s) {
    count_tmp <- data_m[grep(s, names(data_m))]
    if (sorting){
      name_order <- sapply(tool_names, function(t) grep(paste0("__", t, "__"), names(count_tmp)))
      count_tmp <- count_tmp[name_order]
    }
    names(count_tmp) <- tools_nice_m
    
    comb_mat <- make_comb_mat(count_tmp)
    if (get_only_mat){# for suppl fig 2 density plots
      comb_mat
    } else{
      svglite(paste0("plots_revision/tmp/upset/upset_", dataset_name, "_", type_m, "_", s, ".svg"), 
              width = 6, height = 3)
      print(UpSet(comb_mat, 
                  set_order = tools_nice_m, 
                  comb_col = c("#e41a1c",'#377eb8','#4daf4a','#984ea3','#ff7f00')[comb_degree(comb_mat)]))
      dev.off()
      comb_size(comb_mat)
    #comb_mat
    }
  })
}
counts_pbmc_cardiac_endo <- get_counts(pbmc_cardiac_endo)



# make long tables
get_data_upset <- function(data_m, type_m, datasets_nice_m, get_mat){
d <- list()
samples <- c(paste(paste0("sample",seq(4))))
d[[datasets_nice_m[3]]] <- get_upset(samples, data_m$PBMC, type_m=type_m, 
                                     sorting=T, dataset_name="pbmc", get_only_mat=get_mat)
names(d[[datasets_nice_m[3]]]) <- samples

samples <- paste0("MF1701",c(0,3:8))
d[[datasets_nice_m[2]]] <- get_upset(samples, data_m$Cardiac, type_m=type_m, sorting=T, 
                                     dataset_name="cardiac", get_only_mat=get_mat)
names(d[[datasets_nice_m[2]]]) <- samples

samples <- c("Heart","Kidney","Liver","Brain","Colon","Lung","muscle_EDL","muscle_Soleus","Small_Intestine","Spleen","Testis")
d[[datasets_nice_m[1]]] <- get_upset(samples, data_m$Endothelial, type_m=type_m, 
                                     sorting=T, dataset_name="endo", get_only_mat=get_mat)
names(d[[datasets_nice_m[1]]]) <- samples

samples <- paste0("103837-001-00",1:5)
d[[datasets_nice_m[4]]] <- get_upset(samples, data_m$HF, type_m=type_m, sorting=T, 
                                     dataset_name="hf", get_only_mat=get_mat)
names(d[[datasets_nice_m[4]]]) <- samples

d
}
##### make data pretty (for ggplot) #####
transform_data_fig_2 <- function(data_m, datasets_nice_m){
  d <- data.frame(gene_count= unlist(data_m))
  d <- cbind(d, do.call(rbind, strsplit(row.names(d), "\\.")))
  colnames(d) <- c("gene_count", "Dataset", "Samples", "Intersection")

  d_mean <- aggregate(gene_count ~ Dataset + Intersection, data = d, mean)
  d_var <- aggregate(gene_count ~ Dataset + Intersection, data = d, sd)
  d_mean$var <- d_var$gene_count
  colnames(d_mean)[3] <- "mean"
  # sort intersections by number of mapper
  d_mean$Intersection <- factor(d_mean$Intersection)
  cross_total <- apply(do.call(rbind, strsplit(levels(d_mean$Intersection), "")),
                     1, function(x) sum(as.integer(x)))
  names(cross_total) <- seq_along(cross_total)
  cross_total <- sort(cross_total, decreasing = T)
  intersect_ordering <- lapply(unique(cross_total), function(cr){
    ct_pos <- as.integer(names(cross_total[grep(cr,cross_total)]))
    sort(levels(d_mean$Intersection)[ct_pos], decreasing = T)
})

d_mean$Intersection <- factor(d_mean$Intersection, levels = unlist(intersect_ordering)) #levels(plot_data_mean$Intersection)[as.integer(names(cross_total))])
d_mean$Dataset <- factor(d_mean$Dataset, levels = datasets_nice_m)
d_mean
}

make_plot_upset <- function(data_m, type_m){
  ggplot(data_m, aes(x = Dataset, y = mean)) +
    geom_col(aes(fill = Dataset), width = 0.7)  +
    geom_errorbar(aes(x = Dataset, ymin = mean-var, ymax = mean+var, width = 0.4)) +
    #geom_jitter(data=plot_data, aes(x=Dataset, y = gene_count), width = 0.4, size=2, alpha=0.3, show.legend = F) +
    facet_grid(.~Intersection, scales = "free") +
    theme_classic() +
    ylab("Intersection size") +
    ggtitle(label = type_m) +
    #scale_y_continuous(expand = c(0, 0),limits = c(-300,13000)) +
    theme(text = element_text(size = 15), 
          panel.background = element_rect(fill = "white", colour = "grey70"), 
          panel.spacing = unit(0,"mm"), 
          plot.title = element_text(hjust = .5), 
          axis.title.x = element_blank(), 
          axis.text.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          axis.line.x = element_blank(), 
          strip.text.x = element_text(angle = 270))
}


#### get gene counts and prepare for upset plot ####
genes_pbmc_endo_cardiac <- lapply(datasets_nice, function(d) {
  lapply(counts_pbmc_cardiac_endo[grep(d, names(counts_pbmc_cardiac_endo))], function(x) x$genes)
})
names(genes_pbmc_endo_cardiac) <- datasets_nice

comb_sizes_genes <- get_data_upset(genes_pbmc_endo_cardiac, "gene", datasets_nice)
plot_data_genes <- transform_data_fig_2(comb_sizes_genes, datasets_nice)
p_upset_genes <-  make_plot_upset(plot_data_genes, "Intersection of genes")


ggsave("plots_revision/tmp/upset/barplot_gene_hf.svg", 
       p_upset_genes, device = "svg", width = 10, height = 4)



#### Get barcodes per intersection and dataset ####
bc_pbmc_endo_cardiac <- lapply(datasets_nice, function(d) {
  lapply(counts_pbmc_cardiac_endo[grep(d,names(counts_pbmc_cardiac_endo))], function(x) x$cells)
})
names(bc_pbmc_endo_cardiac) <- datasets_nice


comb_sizes_bc <- get_data_upset(bc_pbmc_endo_cardiac, "barcode", datasets_nice)
plot_data_bc <- transform_data_fig_2(comb_sizes_bc, datasets_nice)
p_upset_bc <-  make_plot_upset(plot_data_bc, "Intersection of barcodes")


ggsave("plots_revision/tmp/upset/barplot_barcodes_hf.svg", p_upset_bc, 
       device = "svg", width = 8.5, height = 3.3)



#### Make Density plot for Suppl Fig 2 ####

# complete data
plot_data <- lapply(pbmc_cardiac_endo, function(x) {
  x$nFeature_RNA
})


##### Set up function #####
plot_density <- function(data_m, bc_m, mapper_set, mapper_set_final, dataset_m, set){
  mapper_set <- paste0("__", mapper_set, "__")
  mapper_set_final <- c(mapper_set_final, paste0(mapper_set_final, "*"))
  #make the subsets
  data_m_tmp <- data_m[grep(paste0(mapper_set, dataset_m, collapse = "|"), names(data_m))]
  data_m_tmp_subset <- lapply(data_m_tmp, function(x) {
    x[names(x) %in% bc_m]
  })
  data_names <- unlist(strsplit(names(data_m_tmp_subset),paste0("__", dataset_m)))
  names(data_m_tmp_subset) <- paste0(data_names, paste0("*__", dataset_m))

  #merge whole datasets and subsets together, so that it can be ploted together
  d <- melt(c(data_m_tmp_subset, data_m_tmp))
  d <- cbind(d, do.call(rbind, strsplit(d$L1,"__")))
  colnames(d) <- c("value", "L1", "Samples", "Mapper", "Datasets")
  d$Mapper <- factor(d$Mapper, levels = c(paste0(mapper, "*"),mapper))
  levels(d$Mapper) <- mapper_set_final
  print(str(d))

  ggplot(d, aes(x = value, colour = Mapper)) +
    geom_density(aes(linetype = Mapper), size = 1) +
    scale_x_log10() +
    ggtitle(paste0(set,": ", dataset_m)) +
    theme_article() +
    theme(text = element_text(size = 14), plot.title = element_text(hjust = 0.5)) +
    xlab("Genes per cell (log10)") + ylab("Density")
}

suppl_fig_2 <- list()
comb_sizes_bc <- get_data_upset(bc_pbmc_endo_cardiac, "barcode", 
                                datasets_nice, get_mat = T)

bc_per_sample <- lapply(comb_sizes_bc$Cardiac, function(x) extract_comb(x, "11101"))
bc <- unique(unlist(bc_per_sample))
##### Density for 1:Cardiac #####
bc <- extract_comb(comb_sizes_bc$Cardiac[[1]], "11101")
mapper <- c("cellranger6","starsolo","alevin-fry", "kallisto")
mapper_final <- c("Cell Ranger 6", "STARsolo","Alevin-fry", "Kallisto")
suppl_fig_2[[1]] <- plot_density(plot_data, bc, mapper, mapper_final, "Cardiac","1")
##### Density for 1:HF #####
bc <- extract_comb(comb_sizes_bc$HF[[1]], "11101")
mapper <- c("cellranger6","starsolo","alevin-fry", "kallisto")
mapper_final <- c("Cell Ranger 6", "STARsolo","Alevin-fry", "Kallisto")
suppl_fig_2[[2]] <- plot_density(plot_data, bc, mapper, mapper_final, "HF","1")

##### Density for 2:HF #####
bc <- extract_comb(comb_sizes_bc$Cardiac[[1]], "11001")
mapper <- c("cellranger6","starsolo", "kallisto")
mapper_final <- c("Cell Ranger 6", "STARsolo", "Kallisto")
suppl_fig_2[[3]] <- plot_density(plot_data, bc,mapper, mapper_final, "Cardiac","2")
##### Density for 2:Cardiac #####
bc <- extract_comb(comb_sizes_bc$HF[[2]], "11001")
mapper <- c("cellranger6","starsolo", "kallisto")
mapper_final <- c("Cell Ranger 6", "STARsolo", "Kallisto")
suppl_fig_2[[4]] <- plot_density(plot_data, bc,mapper, mapper_final, "HF","2")

##### Density for 3:HF #####
bc <- extract_comb(comb_sizes_bc$HF[[1]], "00101")
mapper <- c("alevin-fry", "kallisto")
mapper_final <- c("Alevin-fry", "Kallisto")
suppl_fig_2[[5]] <- plot_density(plot_data, bc,mapper, mapper_final, "HF","3")

##### Density for 4:Endothelial #####
bc <- extract_comb(comb_sizes_bc$Endothelial[[1]], "00010")
mapper <- c("alevin")
mapper_final <- c("Alevin")
suppl_fig_2[[6]] <- plot_density(plot_data, bc,mapper, mapper_final, "Endothelial","4")
##### Density for 4:PBMC #####
bc <- extract_comb(comb_sizes_bc$PBMC[[1]], "00010")
mapper <- c("alevin")
mapper_final <- c("Alevin")
suppl_fig_2[[7]] <- plot_density(plot_data, bc,mapper, mapper_final, "PBMC","4")

##### Density for 6:HF #####
bc <- extract_comb(comb_sizes_bc$HF[[1]], "00001")
mapper <- c("kallisto")
mapper_final <- c("Kallisto")
suppl_fig_2[[8]] <- plot_density(plot_data, bc, mapper, mapper_final, "HF","5")

suppl_fig_2[c(2:4,6:8)] <- lapply(suppl_fig_2[c(2:4,6:8)], function(x) x + theme(axis.title.y = element_blank()))
suppl_fig_2[1:4] <- lapply(suppl_fig_2[1:4], function(x) x + theme(axis.title.x = element_blank()))

p_density_all <- ggpubr::ggarrange(plotlist = suppl_fig_2, nrow = 2, ncol = 4,
                                   align = "h")
#p_density_all

ggsave("plots_revision/tmp/suppl_Figure2_densities_hf.svg", p_density_all, device = "svg", width = 16, height = 9)

```


```{r Suppl Figure 3}

filtered <- c("results_filtered_anno", "results_unfiltered_anno")
pbmc_cardiac_endo_id <- save_seurat_objects(datasets, tools, filtered, 1)
names(pbmc_cardiac_endo_id) <- gsub("pbmc_10x","PBMC", names(pbmc_cardiac_endo_id))


all_genes_mouse <- read.table(paste0(github_path, "mapping/post_mapping/all_genes_mouse.tsv"),
                                     header = T,  colClasses = c(rep("character",4)))

all_genes_human <- read.delim(paste0(github_path, "mapping/post_mapping/all_genes_human.tsv"),
                              header = T,  colClasses = c(rep("character",4)))

gene_counts <- lapply(pbmc_cardiac_endo_id, function(x) {
  gene_count <- rowSums(x)
  gene_count <- gene_count[gene_count > 0]
  gene_count
})

pbmc_biotypes <- lapply(gene_counts[grep("PBMC", names(gene_counts))], function(x) {
  if (grepl("\\.",names((x))[1]) ){
    biotype <- all_genes_human[all_genes_human$ID_type %in% names(x),]
  }else {
    biotype <- all_genes_human[all_genes_human$ID %in% names(x),]
  }
  biotype$genetype
})

cardiac_endo_biotypes <- lapply(gene_counts[grep("Cardiac|Endothelial", names(gene_counts))], function(x) {
  if (grepl("\\.",names((x))[1]) ){
    biotype <- all_genes_mouse[all_genes_mouse$ID_type %in% names(x),]
  }else {
    biotype <- all_genes_mouse[all_genes_mouse$ID %in% names(x),]
  }
  biotype$genetype
})

biotypes <- c(pbmc_biotypes, cardiac_endo_biotypes)

biotypes_long <- melt(biotypes)

biotypes_long <- cbind(biotypes_long, do.call(rbind, strsplit(biotypes_long$L1,"__")))
colnames(biotypes_long) <- c("value", "L1", "Sample", "Mapper", "Dataset", "Filtering")
biotypes_long$Dataset <- factor(biotypes_long$Dataset, levels = datasets_nice)
biotypes_long$Mapper <- factor(biotypes_long$Mapper, levels = tools)
levels(biotypes_long$Mapper) <- tools_nice
biotypes_long$Filtering <- factor(biotypes_long$Filtering)
levels(biotypes_long$Filtering) <- c("Filtered", "Unfiltered")
biotypes_long$value <- factor(biotypes_long$value)
biotypes_long$Sample <- factor(biotypes_long$Sample)


b_plot <- biotypes_long %>%
  group_by(Sample, Mapper, Dataset, Filtering, value) %>%
  summarize(n()) %>%
  group_by(Mapper, Dataset, Filtering)
b_plot %>%
  summarise(avg = mean(Mapper, Dataset, Filtering, n()))
  
  
ggplot(b_plot, aes(x = value, y = , color = Mapper, shape = Filtering)) +
  geom_point(position = position_dodge(width = 1), size = 3, alpha = 0.8) +
  #geom_vline(xintercept = seq(levels(biotypes_long$value))+.5, color = "grey90") +
  facet_grid(Dataset~.)# +


# condense biotypes
levels(biotypes_long$value)[grep("TR_", levels(biotypes_long$value))] <- "TR-genes"
levels(biotypes_long$value)[grep("IG_", levels(biotypes_long$value))] <- "IG-genes"
levels(biotypes_long$value)[levels(biotypes_long$value) %in% c("polymorphic_pseudogene",
                                        "pseudogene",
                                        "unitary_pseudogene",
                                        "rRNA_pseudogene"
                                        )] <- "other_pseudogenes"
levels(biotypes_long$value)[levels(biotypes_long$value) %in% c("transcribed_unprocessed_pseudogene",
                                        "transcribed_processed_pseudogene",
                                        "transcribed_unitary_pseudogene")] <- "transcribed_pseudogenes"
levels(biotypes_long$value)[levels(biotypes_long$value) %in% c("translated_unprocessed_pseudogene",
                                        "translated_processed_pseudogene")] <- "translated_pseudogenes"
biotypes_long <- biotypes_long[!biotypes_long$value %in% c("Mt_tRNA","Mt_rRNA","scRNA","sRNA","ribozyme","scaRNA","vaultRNA","snoRNA","snRNA","rRNA"),] # remove unnecessary biotypes

biotypes_long$value <- droplevels(biotypes_long$value)
levels(biotypes_long$value) <- sub("_pseudogene|_pseudogenes", "\npseudogenes", levels(biotypes_long$value))
# change biotypes order
biotypes_long$value <- factor(biotypes_long$value, levels = c("protein_coding",
                                                                   "lncRNA",
                                                                   "IG_genes",
                                                                   "TR_genes",
                                                                   "TEC",
                                                                   "miRNA",
                                                                   "misc_RNA",
                                                                   "processed\npseudogenes",
                                                                   "unprocessed\npseudogenes",
                                                                   "transcribed\npseudogenes",
                                                                   "translated\npseudogenes",
                                                                   "other\npseudogenes"))
#levels(biotypes_long$value)[c(1,3,4,7)] <- c("Protein coding", "IG genes", "TR genes", "misc RNA") 
levels(biotypes_long$value) <- gsub("_", " ", levels(biotypes_long$value))
biotypes_long_long <- melt(table(biotypes_long$value, biotypes_long$Mapper, biotypes_long$Dataset, biotypes_long$Filtering, biotypes_long$Sample))
colnames(biotypes_long_long) <- c("Biotype","Mapper", "Dataset",  "Filtering", "Sample","value")

biotypes_long_long_aggr <- aggregate(value ~ Mapper + Dataset + Filtering + Biotype, data = biotypes_long_long, mean)

biotypes_long_long_pbmc <- biotypes_long_long[biotypes_long_long$Dataset=="PBMC" & 
                                                biotypes_long_long$Sample %in% paste0(c("sample"), c(1:4)),]
biotypes_long_long_cardiac <- biotypes_long_long[biotypes_long_long$Dataset=="Cardiac" &
                                                   biotypes_long_long$Sample %in% paste0(c("MF1701"), c(0,3:8)),]
biotypes_long_long_endo <- biotypes_long_long[biotypes_long_long$Dataset=="Endothelial" &
                                                biotypes_long_long$Sample %in% c("Brain", "Colon", "Heart", "Kidney", "Liver", "Lung", "muscle_EDL", "muscle_Soleus", "Small_Intestine", "Spleen", "Testis"),]
#biotypes_long_long_endo <- biotypes_long_long_endo[biotypes_long_long_endo$Mapper %in% c("Cell Ranger", "STARsolo", "Alevin", "Kallisto"),]

biotypes_long_long2 <- rbind(biotypes_long_long_pbmc, biotypes_long_long_cardiac, biotypes_long_long_endo)
#biotypes_long_long <- biotypes_long_long[biotypes_long_long$value>0,]

biotypes_long_long_aggr <- aggregate(value ~ Mapper + Dataset + Filtering + Biotype, data = biotypes_long_long2, mean)


ggplot(biotypes_long_long_aggr, aes(x = Biotype, y = value, color = Mapper, shape = Filtering)) +
  geom_point(position=position_dodge(width = 1), size = 3, alpha = 0.8) +
  geom_vline(xintercept = seq(levels(biotypes_long_long_aggr$Biotype))+.5, color = "grey90") +
  facet_grid(Dataset~.) +
  scale_shape_manual(values=c(19, 14)) +
  scale_color_brewer(palette = "Set2") +
  scale_y_log10(limits = c(1,20000)) +
  theme_light() +
  guides(shape = guide_legend(title = "Annotation")) +
  theme(text = element_text(size = 17),
        axis.text.x = element_text(angle = 60, vjust = +1, hjust = +1.1),
        panel.grid = element_blank())


ggsave("plots_revision/Suppl_figure_3_log10.svg", device = "svg", width = 13, height = 7)

```


```{r Figure 3 olfr and vmn gene family}
#### Set up functions ####
get_counts_vmn_olfr <- function(data_m, organism)
cardiac_genes <- lapply(data_m, function(x) {
  if (organism == "mouse"){
    olfr_tmp = rowSums(x[grep("^Olfr", row.names(x))])
    vmn_tmp = rowSums(x[grep("^Vmn", row.names(x))])
  } else if(organism == "human"){
    olfr_tmp = rowSums(x[grep("^OR[0-9]{1}[A-Z]{1}[0-9]+", row.names(x))])
    vmn_tmp = rowSums(x[grep("^VN[0-9]{1}[A-Z]{1}[0-9]+", row.names(x))])
  }
  olfr_tmp <- olfr_tmp[olfr_tmp>0]
  vmn_tmp <- vmn_tmp[vmn_tmp>0]
  list(olfr = olfr_tmp, vmn = vmn_tmp)
})

prepare_data <- function(data_m, tools_m, tools_nice_m, datasets_nice_m){
  d <- melt(data_m)
  d <- cbind(d, do.call(rbind, strsplit(d$L1, "__")))
  colnames(d) <- c("value", "L1", "Samples", "Mapper", "Dataset")
  
  samples_endo <- c("Brain", "Colon", "Heart", "Kidney", "Liver", "Lung", "muscle_EDL", 
                    "muscle_Soleus", "Small_Intestine", "Spleen", "Testis")
  d$Samples <- factor(d$Samples, levels = c(samples_endo, paste0("MF1701",c(0,3:8)),
                                                            paste0("sample",1:4), paste0("103837-001-00",1:5)))
  levels(d$Samples) <- c(samples_endo, "Control", paste0("Day ",c(1,3,5,7,14,28)), 
                                 paste0("N",1:4), paste0("Patient ",1:5))

  d$Mapper <- factor(d$Mapper, levels = tools_m)
  levels(d$Mapper) <- tools_nice_m
  d$Dataset <- factor(d$Dataset, levels = datasets_nice_m)
  d
}

cardiac_endo_genes_vmn_olfr <- get_counts_vmn_olfr(pbmc_cardiac_endo[grep("PBMC|HF",names(pbmc_cardiac_endo), invert = T)], "mouse")
pbmc_genes_vmn_olfr <- get_counts_vmn_olfr(pbmc_cardiac_endo[grep("PBMC|HF",names(pbmc_cardiac_endo), invert = F)], "human")

genes_vmn_olfr <- c(cardiac_endo_genes_vmn_olfr, pbmc_genes_vmn_olfr)

#### plot histograms ####
figure3_list <- list()
plot_gene_families <- function(plotdata_umi, plotdata_gene, title_m){
ggplot(plotdata_umi, aes(x = Samples, y = value, fill = Dataset)) +
  geom_boxplot() + 
  geom_line(data = plotdata_gene, aes(group = 1, colour = "red"), size = 1, show.legend = T) +
  facet_grid(.~Mapper, scales = "free", space = "free") +
  theme_light() +
  coord_cartesian(ylim = c(1.68, 11000)) +
  annotation_logticks(sides = "l") +
  xlab("") +
  ylab("UMI-count per gene [log10]") +
  ggtitle(title_m) +
  scale_y_log10() +
  theme(text = element_text(size = 15), 
        axis.text.x = element_text(angle = 60, vjust=+1.05, hjust = +1.1), 
        panel.grid.major.x = element_blank(), 
        axis.text.x.bottom = element_text(size = 10), 
        plot.title = element_text(hjust = 0.5)) 
}

##### Olfr #####
umicount_olfr <- lapply(genes_vmn_olfr, function(x) x$olfr)
gene_count_olfr <- lapply(genes_vmn_olfr, function(x) length(x$olfr))

plot_data_umi_count <- prepare_data(umicount_olfr, tools, tools_nice, datasets_nice)
plot_data_gene_count <- prepare_data(gene_count_olfr, tools, tools_nice, datasets_nice)

figure3_list[["olfr"]] <- plot_gene_families(plot_data_umi_count, plot_data_gene_count, "Olfr")

##### Vmn #####
umicount_vmn <- lapply(genes_vmn_olfr, function(x) x$vmn)
gene_count_vmn <- lapply(genes_vmn_olfr, function(x) length(x$vmn))

plot_data_umi_count <- prepare_data(umicount_vmn, tools, tools_nice, datasets_nice)
plot_data_gene_count <- prepare_data(gene_count_vmn, tools, tools_nice, datasets_nice)

figure3_list[["vmn"]] <- plot_gene_families(plot_data_umi_count, plot_data_gene_count, "Vmn")

#### save figure ####
figure3 <- ggpubr::ggarrange(plotlist = figure3_list, ncol = 1, nrow = 2,
                             labels = "AUTO", common.legend = T, legend = "bottom")

ggsave("plots_revision/Figure_3_hf.svg", figure3, device = "svg", width = 14, height = 13)

figure3
```


```{r Integration}
#### Set up function ####

integrate_per_mapper <- function(data_m, project_name, dims, resolution){
  seurat_features <- SelectIntegrationFeatures(object.list = data_m, nfeatures = 2500)
  data_m <- PrepSCTIntegration(object.list = data_m, anchor.features = seurat_features)

  seurat_anchors <- FindIntegrationAnchors(object.list = data_m, normalization.method = "SCT", 
                                           anchor.features = seurat_features)
  seurat_integrated <- IntegrateData(anchorset = seurat_anchors, normalization.method = "SCT")

  seurat_integrated <- RunPCA(seurat_integrated)
  print(ElbowPlot(seurat_integrated))
  seurat_integrated <- RunUMAP(seurat_integrated, dims = 1:dims)
  seurat_integrated <- FindNeighbors(seurat_integrated, dims = 1:dims)
  #set_resolution <- resolution
  seurat_integrated <- FindClusters(seurat_integrated, resolution = resolution)
  Project(seurat_integrated) <- project_name
  seurat_integrated
}

#### Integration ####

pbmc_cardiac_endo <- lapply(pbmc_cardiac_endo, function(x) {
  x <- RenameCells(x, add.cell.id = strsplit(Project(x),"__")[[1]][1])
  subset(x, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10)
})

pbmc_cardiac_endo <- lapply(pbmc_cardiac_endo, function(x) {
  print(Project(x))
  x <- SCTransform(x, verbose = T)
  })


integrated_data <- list()
integrated_data[["PBMC"]] <- lapply(tools, function(x) {
  integrate_per_mapper(pbmc_cardiac_endo[grep(paste0(x,"__PBMC"), names(pbmc_cardiac_endo))], 
                       paste("PBMC_integration", x, sep = "_"), 20, 0.15)
})


integrated_data[["Cardiac"]] <- lapply(tools, function(x) {
  integrate_per_mapper(pbmc_cardiac_endo[grep(paste0(x,"__Cardiac"), names(pbmc_cardiac_endo))], 
                       paste("Cardiac_integration", x, sep = "_"), 20, 0.15)
})


integrated_data[["Endothelial"]] <- lapply(tools, function(x) {
  integrate_per_mapper(pbmc_cardiac_endo[grep(paste0(x,"__Endothelial"), names(pbmc_cardiac_endo))], 
                       paste("Endothelial_integration", x, sep = "_"), 20, 0.15)
})

data_name <- "HF"
integrated_data[[data_name]] <- lapply(tools, function(x) {
  integrate_per_mapper(pbmc_cardiac_endo[grep(paste0(x, "__", data_name), names(pbmc_cardiac_endo))], 
                       paste(data_name, "integration", x, sep = "_"), 20, 0.15)
})

names(integrated_data$PBMC) <- tools_nice
names(integrated_data$Cardiac) <- tools_nice
names(integrated_data$HF) <- tools_nice
integrated_data$PBMC <- lapply(integrated_data$PBMC, function(x) {
  DefaultAssay(x) <- "SCT"
  x
  })
for (i in integrated_data$HF) DefaultAssay(i) <- "integrated"
lapply(integrated_data$HF, function(x) {
  DimPlot(FindClusters(x,resolution = 0.5), label = T)
})



DimPlot(integrated_data$PBMC$STARsolo, label = T)

lapply(integrated_data, function(x) {
  lapply(x, function(p) {
      DimPlot(p, label = T)
      ggsave(paste0(main_outpath, "plots_revision/tmp/umap/umap_", Project(p), ".png"))
  })
})


integrated_data$PBMC <- lapply(integrated_data$PBMC, function(x) {
  x <- RunPCA(x, verbose = FALSE)
  x <- RunUMAP(x, dims = 1:20, verbose = FALSE)
  x <- FindNeighbors(x, dims = 1:20, verbose = FALSE)
  x <- FindClusters(x, resolution = 0.4) 
  x
})

DotPlot(integrated_data$PBMC$`Alevin-fry`, features = list("T-cell CD4+" = c("IL7R", "CCR7"), "CD14+ Mono" = c("CD14", "LYZ"), "Memory CD14+" = c("S100A4"), "B" = c("MS4A1"), "T-cell CD8+" = c("CD8A"),"FCGR3A+ Mono" = c("FCGR3A","MS4A7"), "NK" = c("GNLY","NKG7"), "DC" = c("FCER1A","CST3"), "Platelets" = c("PPBP")), assay = "SCT")

FeaturePlot(integrated_data$PBMC$STARsolo, features = c("IL7R", "CCR7", "CD14", "LYZ", "S100A4", "MS4A1", "CD8A", "FCGR3A","MS4A7","GNLY","NKG7","FCER1A","CST3","PPBP"), order = T)

VlnPlot(integrated_data$PBMC$STARsolo, features = c("IL7R", "CCR7", "CD14", "LYZ", "S100A4", "MS4A1", "CD8A", "FCGR3A","MS4A7","GNLY","NKG7","FCER1A","CST3","PPBP"))


VlnPlot(integrated_data$PBMC$STARsolo, features = c("CD8A", "GZMK", "CCL5", "CCR7","IL32", "ISG15"),assay = "RNA")

```


```{r}
f <- read.table("/media/Helios_scStorage/Ralf/triangulate_two_level/harmonizing/PanglaoDB_markers_27_Mar_2020.tsv", header = T, sep = "\t", quote = '')

fheart <- f[grepl("Hs",f$species) & f$organ %in% c("Blood", "Connective tissue", "Epithelium","Immune system","Heart","Smooth muscle", "Vasculature"),]

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}

cell_types <- unique(fheart$cell.type)
names(cell_types) <- cell_types

marker_genes <- lapply(cell_types, function(c){
  unique(toupper(as.character(fheart[fheart$cell.type==c,2])))
})
marker_genes <- marker_genes[-c(49)]
for (m in seq_along(marker_genes)){
  p_marker<- DotPlot(integrated_data$HF$`Cell Ranger 6`, features = marker_genes[[m]], assay = "RNA") +
    theme(axis.text.x = element_text(angle = 60,vjust=+1.05 ,hjust = +1.1))
  ggsave(paste0("plots_revision/tmp/marker_genes/HF/", names(marker_genes)[m] , ".svg"), p_marker, device = "svg")
}
DotPlot(seurat_object, features = marker_genes[1:5])

seurat_cluster <- seurat_object$seurat_clusters
levels(seurat_cluster) <- c("FB","CM","Monocytes", "EC","EC", "Mural cells", "Pericardium-like", "Lymphocytes", "Adipocyte-like", "CM", "EC-like", "ECFB", "CM")
seurat_object$Celltypes <- seurat_cluster
```


```{r SCINA cell annotation}
#library(SCINA)

# Old
# cell_type_genes_pbmc <- list(
#   "Endothelial cells"=c("Ly6c1","Egfl7","Gpihbp1","Cdh5","Mgll","Slc9a3r2","Emcn","Kdr","Pecam1","Rgcc","Tie1","Cldn5","Eng"),
#   Monocytes=c("Dab2","Adgre1","Mrc1","Hpgd","P2ry6","C3ar1","F13a1","Maf","Ms4a7","Cd14","Ptprc","Lyz","Fcgr3a"),
#   "Dendritic cells"=c("Fcer1a", "Cst3", "Il3ra", "Lilra4", "Dnase1l3", "Rnase6"),
#   "NK-cells"=c("Gnly","Nkg7","Klrk1","Klre1","Klrd1","Ncr1","Ctsw","Klrb1c","Gzma","Gzmb","Prf1"),
#   "B-cells"=c("Cd79a","Ly6d","Cd79b","H2-DMb2","Ms4a1","H2-Ob","Fcmr","Ccr7","Bank1","Cd55","Iglc2", "Iglc3","Cd19","Pax5","Iglc1"),
#   "T-cell CD4+"=c("Cd3g","Lat","Cd3e","Il7r","Cd247","Itk","Icos","Trdc"),
#   "T-cell CD8+"=c("Cd8a","Cd8b","Gzmk", "Ccl5","cd27","Themis")
# )

#### Define cell type marker ####
cell_type_genes_pbmc <- list(
  "Platelet" = c("PPBP","ITGA2B", "PF4", "GP1BA", "SELP", "TUBB1", "ITGB3", 
                 "NCOA4", "RGS18", "TREML1", "CD9", "NRGN", "MYL9"),
  Monocytes = c("Ms4a7","Cd14","Ptprc","Lyz","Fcgr3a"),
  "Dendritic cells" = c("Fcer1a", "Cst3", "Il3ra", "Lilra4", "Dnase1l3", "Rnase6"),
  "NK-cells" = c("Gnly","Nkg7","Klrk1","Klrd1","Ncr1","Ctsw","Gzma","Gzmb","Prf1"),
  "B-cells" = c("Cd79a","Ly6d","Cd79b","Ms4a1","Fcmr","Bank1",
              "Cd55","Iglc2", "Iglc3","Cd19","Pax5","Iglc1"),
  "T-cell CD4+" = c("CD4","Cd3g","Lat","Cd3e","Il7r","Cd247","Itk","Icos","Trdc","Ccr7"),
  "T-cell CD8+" = c("Cd8a","Cd8b","cd27","Themis")#"Gzmk", "Ccl5",
)
cell_type_genes_pbmc <- sapply(cell_type_genes_pbmc,toupper)

gene2celltype_pbmc <- reshape2::melt(cell_type_genes_pbmc)
gene2celltype_pbmc$value <- as.character(gene2celltype_pbmc$value)
colnames(gene2celltype_pbmc) <- c("gene","cell_type")

cell_type_genes_cardiac <- list(
  Fibroblasts=c("Lamc1","Pcsk6","Pdgfra","Entpd2","Dpep1","Adamts5","Medag","Ms4a4d","Lamb1","Tcf21","Col1a1","Dkk3","Tbx20","Wif1","Frzb","Meox1","Prg4","Abi3bp","Pdgfrl","Mdk","Gstm5"),
  "Endothelial cells"=c("Ly6c1","Egfl7","Gpihbp1","Cdh5","Mgll","Slc9a3r2","Emcn","Kdr","Pecam1","Rgcc","Tie1","Cldn5","Eng"),
  Granulocytes=c("S100a8","S100a9","Slpi","Csf3r","Hcar2","Lmnb1","Retnlg","Clec4d","Hp","Hdc"),
  Monocytes=c("Dab2","Adgre1","Mgl2","Mrc1","Hpgd","P2ry6","C3ar1","F13a1","Maf","Ms4a7","Cd14","Fcgr1","Ptprc","Lgals3","Napsa","Plbd1","Ccr2","Rnase6","Plac8","Ifitm6","Naaa","Ear2","H2afy"),
  "NKT-cells"=c("Ccl5","Nkg7","Klrk1","Klre1","Klrd1","Ncr1","Ctsw","Klrb1c","Gzma","Gzmb","Cd3d","Skap1","Lef1","Tcf7"),
  "B-cells"=c("Cd79a","Ly6d","Cd79b","H2-DMb2","Ms4a1","H2-Ob","Fcmr","Ccr7","Bank1","Cd55","Iglc2", "Iglc3","Cd19","Pax5","Iglc1"),
  "T-cells"=c("Cd3g","Lat","Cd3e","Il7r","Cd247","Itk","Icos","Trdc"),
  "Mural cells"=c("Kcnj8","Vtn","Colec11","Steap4","Abcc9","Myo1b","Cog7","P2ry14","Heyl","Gnb4","Tagln","Mustn1","Myh11","Mylk","Pcp4l1","Sncg","Lmod1","Des","Pln","Nrip2")
)
gene2celltype_cardiac <- reshape2::melt(cell_type_genes_cardiac)
gene2celltype_cardiac$value <- as.character(gene2celltype_cardiac$value)
colnames(gene2celltype_cardiac) <- c("gene","cell_type")



lapply(integrated_data$PBMC, function(x) {
      DotPlot(x, features = cell_type_genes_pbmc, assay = "RNA") +
              theme(axis.text.x = element_text(angle = 60,vjust=+1.05 ,hjust = +1.1))
      ggsave(paste0("plots_revision/tmp/marker_genes/Marker_genes_", Project(x), "2.png"),
             height = 6, width = 15)
})

lapply(integrated_data$Cardiac, function(x) {
      DotPlot(x, features = cell_type_genes_cardiac, assay = "RNA") +
              theme(axis.text.x = element_text(angle = 60,vjust=+1.05 ,hjust = +1.1))
      ggsave(paste0("plots_revision/tmp/marker_genes/Marker_genes_", Project(x), ".png")
             ,height = 6, width = 15)
})


cell_type_genes_HF <- list(CM=c("TTN","TNNT2","DMD","RYR2","TNNI3","MYH7"),
                   EC=c("CDH5","PECAM1","VWF","ICAM1","CD93","ADGRL4","ENG","EMCN","TIE1"),
                   FB=c("PDGFRA","COL1A1","COL1A2","DCN","DDR2","ADAMTS5","DPT","VCAN","LAMB1","MMP2"),
                   PC=c("PDGFRB","KCNJ8","STEAP4","ABCC9"),
                   SMC=c("ACTA2","TAGLN","MYH11"),
                   Leuko=c("CD163","CD74","LYZ","CD14","PTPRC"),
                   NLC=c("NRXN1","NCAM1","NRXN3","ERBB3"))

gene2celltype_HF <- reshape2::melt(cell_type_genes_HF)
gene2celltype_HF$value <- as.character(gene2celltype_HF$value)
colnames(gene2celltype_HF) <- c("gene","cell_type")


lapply(integrated_data$HF, function(x) {
      DotPlot(x, features = cell_type_genes_HF, assay = "RNA") +
              theme(axis.text.x = element_text(angle = 60,vjust=+1.05 ,hjust = +1.1))
      ggsave(paste0("plots_revision/tmp/marker_genes/Marker_genes_", Project(x), ".png"),
             height = 6, width = 15)
})

```



```{r Suppl Figure 6 and 7 Heatmap}
#library(SCINA)

#### Set up functions ####
jaccard <- function(set1,set2){
  return(length(intersect(set1,set2))/length(union(set1,set2)))
}

precision <- function(set1, set2){
  bool_table <- table(set2 %in% set1)
  #print(bool_table)
  if (length(bool_table) == 1){
    ifelse(names(bool_table)=="FALSE", return(0), return(1))
    # if (names(bool_table)=="FALSE"){
    #   return(1)
    # }else {return(0)}
  }else{
    return(bool_table[["TRUE"]]/sum(bool_table))
  }
}

scina_classi <- function(data_m, marker){
  d <- lapply(data_m, function(x) SCINA(as.data.frame(x@assays$SCT@data), marker))
  lapply(d, function(x) {
    names(x$cell_labels) <- colnames(x$probabilities)
    x$cell_labels
  })
}
## Merge the integrated data sets for the barcode, so that we get the cell type annotation per barcodes across all tools.
merge_list <- function(data_m){
  d <- data.frame(Cellranger_cluster=data_m[[1]], stringsAsFactors = F)
  d$Row.names <- row.names(d)
  for (i in 2:length(data_m)){
    d <- merge(data.frame(data_m[[i]], stringsAsFactors = F), 
               d, 
               by.y="Row.names", 
               by.x=0, 
               all=T, 
               suffixes = names(data_m)[i], ".y")
    colnames(d)[2] <- paste0(names(data_m)[i], "_cluster")
  }
  return(d)
}

major_vote <- function(data_m){
  results <- data.frame(matrix(ncol = 5, nrow = nrow(data_m)), row.names = data_m$Row.names,stringsAsFactors = F)
  
  for (i in 1:nrow(data_m)){
    occurence <- table(as.character(data_m[i,2:ncol(data_m)]),useNA = "always")
    cell_type_proportion <- unlist(lapply(occurence,function(x) x / (ncol(data_m)-1) ))
    cell_type_proportion <- sort(cell_type_proportion,decreasing = T)
  
    if (length(cell_type_proportion) > 1 && is.na(names(cell_type_proportion[1])) ){
      primary_output <- c(cell_type_proportion[2], sort(occurence,decreasing = T)[2], names(cell_type_proportion[2]))
      secondary_output <- c(cell_type_proportion[1], names(cell_type_proportion)[1])
      
    }else if (length(cell_type_proportion)>1) {
      primary_output <- c(cell_type_proportion[1], occurence[1], names(cell_type_proportion[1]))
      secondary_output <- c(cell_type_proportion[2], names(cell_type_proportion)[2])
    }
    results[data_m[i,1],] <- unname(c(primary_output,secondary_output))
  }
  colnames(results) <- c("prop_prim_type", "n_prim_type", "primtype", "prop_sec_type", "sec_type")
  return(results)
}



## just checking if alevin unique bc from upset plots are appearing in the majority voting when bc appearing in more than 2. Obviously not.
#for (i in bc_unique_alevin){
#  bc_rownames <- matrix(unlist(strsplit(row.names(barcode_confident_cell_type),"_")),ncol = 2,byrow = T)[,2]
#  print(str(bc_rownames[bc_rownames %in% i]))
#}

get_scina_bc <- function(cell_type_marker, confident_celltypes){
  barcodes_scina <- lapply(c(names(cell_type_marker), "unknown"),function(x) {
    row.names(confident_celltypes[confident_celltypes$primtype %in% x,])
  })
  names(barcodes_scina) <- c(names(cell_type_marker), "unknown")
  return(barcodes_scina)
}

get_seurat_bc <- function(integrated_data){
  lapply(integrated_data, function(x) {
    barcodes <- split(colnames(x), x$seurat_clusters)
    barcodes
  })
}

##### Plotting #####
# Prepare data for overlap calculation
get_plot_data <- function(cell_type_marker, confident_celltypes, integrated_data, 
                          tools_m=tools, tools_nice_m=tools_nice, metric){

  barcodes_scina <- get_scina_bc(cell_type_marker, confident_celltypes)
  
  barcodes_seurat <- get_seurat_bc(integrated_data)
  
  barcodes_seurat <- unlist(barcodes_seurat, recursive = F)
  n_cluster_seurat <- length(barcodes_seurat)
  n_cluster_scina <- length(barcodes_scina)
  k=1
  # Save all indexes in results to create plot
  results <- data.frame(matrix(ncol = 3, nrow = n_cluster_seurat*n_cluster_scina))
  ## Calculate the overlap with the jaccard index
  for (i in 1:n_cluster_scina){
    for (j in 1:n_cluster_seurat){
      if (metric == "jaccard"){
        indicies <- jaccard(barcodes_scina[[i]], barcodes_seurat[[j]])
      }else if (metric == "precision"){
        indicies <- precision(barcodes_scina[[i]], barcodes_seurat[[j]])
      }
      
      results[k,] <- c(names(barcodes_scina)[i], names(barcodes_seurat)[j], indicies)
      k <- k+1
    }
  }

  ## Prepare data.frame for ggplot
  order_after_cluster_size <- names(sort(sapply(barcodes_scina, length), decreasing = T))
  # move unknown "celltype" at the end of cell types
  pos_unknown <- grep("unknown",order_after_cluster_size)
  order_after_cluster_size <- c(order_after_cluster_size[-c(pos_unknown)],"unknown")
  #order_after_cluster_size <- order_after_cluster_size[c(5,1,2,4,5,6,7,8,9,3)]
  results <- cbind(results, matrix(unlist(strsplit(results$X2, "[.]")), ncol = 2, byrow = T))
  results <- data.frame(SCINA_cell_types = factor(results$X1, levels = order_after_cluster_size),
                        Seurat_clusters = results$X2,
                        Barcode_overlap = as.numeric(results$X3),
                        Tool = factor(results$"1", levels = tools_nice_m),
                        Cluster = results$`2`, stringsAsFactors = T)
  results$Cluster <- factor(results$Cluster, levels = as.character(0:(length(levels(results$Cluster))-1)))
  #print(str(results))
  #levels(results$Tool) <- tools_nice_m
  
  return(results)
}


plot_heatmap <- function(data_m, title_name){
  heatmap_barcodes <- ggplot(data_m, aes(x = SCINA_cell_types, y = Cluster)) +
    geom_tile(aes(fill = Barcode_overlap)) +
    scale_fill_gradientn(colours = c("#4575b4","#ffffbf","#d73027")) +
    facet_grid(Tool~., scales = "free", space = "free") +
    ggtitle(title_name) +
    xlab("SCINA cell type") + 
    ylab("Seurat cluster") +
    theme(text = element_text(size = 20),
          plot.title = element_text(hjust = .5),
          axis.text.x = element_text(angle = 60, vjust= +1.05, hjust = +1.1),
          panel.background = element_blank(),
          panel.spacing = unit(0, "mm"))

  heatmap_barcodes$labels$fill <- "Barcode\noverlap"

  return(heatmap_barcodes)
}

#### Create plots ####
figure6 <- list()

#### PBMC ####
scina_pbmc <- scina_classi(integrated_data$PBMC, cell_type_genes_pbmc)

barcode_ident <- merge_list(scina_pbmc)
confident_cell_type_pbmc_all <- major_vote(barcode_ident)
confident_cell_type_pbmc <- confident_cell_type_pbmc_all[confident_cell_type_pbmc_all$n_prim_type >= 2,] # For Suppl. Figure 5
results_pbmc <- get_plot_data(cell_type_genes_pbmc, confident_cell_type_pbmc, 
                              integrated_data$PBMC, metric = "jaccard")
figure6$pbmc <- plot_heatmap(results_pbmc, "PBMC")

#### Cardiac ####

scina_cardiac <- scina_classi(integrated_data$Cardiac, cell_type_genes_cardiac)
barcode_ident <- merge_list(scina_cardiac)
confident_cell_type_cardiac_all <- major_vote(barcode_ident)
confident_cell_type_cardiac <- confident_cell_type_cardiac_all[confident_cell_type_cardiac_all$n_prim_type >= 2,] # For Suppl. Figure 5 ,7
results_cardiac <- get_plot_data(cell_type_genes_cardiac, confident_cell_type_cardiac, 
                                 integrated_data$Cardiac,  metric = "jaccard")
figure6$cardiac <- plot_heatmap(results_cardiac, "Cardiac")
figure6$cardiac <- figure6$cardiac + ylab("")

ggpubr::ggarrange(plotlist = figure6, labels = "AUTO", font.label = list(size = 16), 
                  legend = "right", common.legend = T)

ggsave("plots_revision/Suppl_figure_6_jaccard.svg", device = "svg", height = 14, width = 14)



#### Suppl. Figure 7 ####
make_collapsed_heatmap <- function(data_m, title_name, tools_nice_m=tools_nice){
  heatmap_metric <- data.frame(data_m %>% group_by(Tool))
  heatmap_metric <- heatmap_metric[heatmap_metric$Barcode_overlap > 0,]
  colapsed_cluster_values <- aggregate(Barcode_overlap ~ SCINA_cell_types + Tool, heatmap_metric, mean)
  colapsed_cluster_values <- colapsed_cluster_values[colapsed_cluster_values$SCINA_cell_types!="unknown",]

  p <- ggplot(colapsed_cluster_values, aes(x = SCINA_cell_types, y = factor(Tool,levels = rev(tools_nice_m)))) +
    geom_tile(aes(fill = signif(Barcode_overlap, 3)*100)) +
    geom_text(aes(label = signif(Barcode_overlap, 3)*100), color = "grey95") + ylab("Tools") +
    scale_y_discrete(expand = c(0, 0)) +
    scale_x_discrete(expand = c(0, 0)) +
    theme_classic() +
    ggtitle(title_name) +
    xlab("SCINA cell types") +
    theme(text = element_text(size = 14), 
          axis.text.x = element_text(angle = 60, vjust = +1.05 , hjust = +1.1),
          axis.line = element_blank(),
          plot.title = element_text(hjust = .5))
  p$labels$fill <- "Barcode\noverlap"
  return(p)
}

figure7 <- list()
figure7$pbmc <- make_collapsed_heatmap(results_pbmc, "PBMC")
figure7$pbmc <- figure7$pbmc + xlab("")
figure7$cardiac <- make_collapsed_heatmap(results_cardiac, "Cardiac")

ggpubr::ggarrange(plotlist = figure7, labels = "AUTO", nrow = 2, ncol = 1)
ggsave("plots_revision/Suppl_figure_7_precision_0.svg", device = "svg", height = 10, width = 8)





# get mean overlap per tool
#results$tools <- matrix(unlist(strsplit(as.character(results$Seurat_clusters),"[.]"),recursive = F),ncol = 2,byrow = T)[,1]

#heatmap_metric <- results %>% group_by(Tool) %>% filter(Barcode_overlap>0.05) %>% summarise(mean = mean(Barcode_overlap))
overlap_per_tool_cell_type <- data.frame(results %>% group_by(Tool,SCINA_cell_types) %>% summarise(sum = sum(Barcode_overlap))) #When looking for the entire overlap per tool
#heatmap_metric <- data.frame(results %>% group_by(Tool) %>% filter(Barcode_overlap>0.05))
#heatmap_metric <- data.frame(heatmap_metric)[c(4:7,9:10),]
heatmap_metric <- data.frame(results %>% group_by(Tool))
heatmap_metric <- heatmap_metric[heatmap_metric$Barcode_overlap>0,]
colapsed_cluster_values <- aggregate(Barcode_overlap ~ SCINA_cell_types + Tool,heatmap_metric,sum)
colapsed_cluster_values <- colapsed_cluster_values[colapsed_cluster_values$SCINA_cell_types!="unknown",]

ggplot(colapsed_cluster_values, aes(x=SCINA_cell_types,y=factor(Tool,levels = c("Kallisto","Alevin","STARsolo","Cell Ranger 5","Cell Ranger")))) +
  geom_tile(aes(fill = Barcode_overlap)) + 
  geom_text(aes(label = round(Barcode_overlap, 3)*100)) + ylab("Tools") + 
  theme(axis.text.x = element_text(angle = 60, vjust=+1.05 , hjust = +1.1))

aggregate(Barcode_overlap ~ Tool,colapsed_cluster_values,mean)
#         Tool Barcode_overlap
#1 Cell Ranger       0.7464897
#2    STARsolo       0.7321907
#3      Alevin       0.6990806
#4    Kallisto       0.6673051



#### Precision ####
scina_celltyes <- get_scina_bc(cell_type_genes_pbmc, confident_cell_type_pbmc)
seurat_melt <- lapply(test_seurat, function(x) {
    unlist(x,use.names = F)
  })
global_overlap <- list()
for (i in seq(scina_celltyes)){
  for (j in seq(seurat_melt)){
    global_overlap[[paste(names(scina_celltyes)[i], 
                          names(seurat_melt)[j], 
                          sep = "__")]] <- precision(scina_celltyes[[i]], 
                                                     seurat_melt[[j]])
  }
}
global_overlap_melt <- melt(global_overlap)
global_overlap_melt <- cbind(global_overlap_melt, 
                             do.call(rbind, strsplit(global_overlap_melt$L1, "__")))
colnames(global_overlap_melt) <- c("overlap", "L1", "celltypes", "Tools")


ggplot(global_overlap_melt, aes(x = celltypes, y = factor(Tools,levels = rev(tools_nice)))) +
    geom_tile(aes(fill = overlap)) +
    geom_text(aes(label = signif(overlap, 3)*100), color = "grey95") + ylab("Tools") +
    scale_y_discrete(expand = c(0, 0)) +
    scale_x_discrete(expand = c(0, 0)) +
    theme_classic() +
    ggtitle("PBMC") +
    xlab("SCINA cell types") +
    theme(text = element_text(size = 14), 
          axis.text.x = element_text(angle = 60, vjust = +1.05 , hjust = +1.1),
          axis.line = element_blank(),
          plot.title = element_text(hjust = .5))
```



```{r Precision and Recall}

#### Set up functions ####
cluster2celltype <- function(data_m, tool_name){
  max_overlap<- data_m %>% filter(Tool == tool_name) %>% group_by(Cluster) %>% slice_max(Barcode_overlap)
  duplets <- duplicated(max_overlap$Cluster)
  if (any(duplets)){
    if (all(max_overlap$Barcode_overlap[duplets]==0)){
      max_overlap <- max_overlap[c(1:(anyDuplicated(max_overlap$Cluster)-2),nrow(max_overlap)),]
    }
  }
  new_vec <- max_overlap$SCINA_cell_types
  names(new_vec) <- max_overlap$Cluster
  new_vec
}
# new_clus_id<- list(cellranger6_celltypes = cluster2celltype(results_pbmc, "Cell Ranger 6"),
#                    starsolo_celltypes = cluster2celltype(results_pbmc, "STARsolo"),
#                    alevin_celltypes = cluster2celltype(results_pbmc, "Alevin"),
#                    alevin_fry_celltypes = cluster2celltype(results_pbmc, "Alevin-fry"),
#                    kallisto_celltypes = cluster2celltype(results_pbmc, "Kallisto"))
# 
# 
# 
# for (i in seq(integrated_data$PBMC)){
#   integrated_data$PBMC[[i]]$celltypes <- integrated_data$PBMC[[i]]$seurat_clusters
#   levels(integrated_data$PBMC[[i]]$celltypes) <- new_clus_id[[i]]
# }


get_category <- function(data_m, confident_celltypes) {
  lapply(data_m, function(x){
    bc_celltypes <- FetchData(x, "celltypes")
    class_comparison <- merge.data.frame(bc_celltypes, confident_celltypes, 
                                         by = 0, all = T)
    class_comparison <- class_comparison[c("Row.names", "celltypes", "primtype")]
    
    class_comparison <- class_comparison %>% count(celltypes, primtype)
    class_comparison$type <- NA
    
    for (i in seq(nrow(class_comparison))){
      if (!is.na(class_comparison[[i,1]]) & !is.na(class_comparison[[i,2]])){
        if(class_comparison[[i,1]] == class_comparison[[i,2]]){
          class_comparison[[i,4]] <- "TP"
        }else {
        class_comparison[[i,4]] <- "FP"
      }
      }else if(is.na(class_comparison[[i,1]]) & !is.na(class_comparison[[i,2]])){
        class_comparison[[i,4]] <- "FN"
      }else if(!is.na(class_comparison[[i,1]]) & is.na(class_comparison[[i,2]])){
        class_comparison[[i,4]] <- "FP"
      }
    }
    class_comparison
  })
}

get_metrics <- function(data_m){
  lapply(data_m, function(x) {
  precision <- list()
  recall <- list()
  F1 <- list()
  for (i in levels(droplevels(x$celltypes))){#
    #print(levels(droplevels(x$celltypes)))
    #print(unique(x$celltypes))

    d_tmp <- x %>% filter(celltypes==i)
    #print(d_tmp)
    TP <- d_tmp %>% filter(type =="TP") %>% pull(n) %>% sum()
    #cat("TP: ")
    #print(d_tmp %>% filter(type =="TP"))
    #print(TP)
    #sum(d_tmp[d_tmp$type=="TP","n"])
    FP <- d_tmp %>% filter(type =="FP") %>% pull(n) %>% sum()
    # cat("FP: ")
    # print(d_tmp %>% filter(type =="FP"))
    # print(FP)
    #FP <- sum(d_tmp[d_tmp$type=="FP","n"])
    FN <- x %>% filter(is.na(celltypes), primtype == i) %>% pull(n) %>% sum()
    # cat("FN: ")
    # print(x %>% filter(is.na(celltypes), primtype == i))
    # print(FN)
    precision[[i]] <- TP / (TP + FP)
    recall[[i]] <- TP / (TP + FN)
    F1[[i]] <- 2 * (precision[[i]] * recall[[i]]) / (precision[[i]] + recall[[i]])
  }
  list(precision = precision, recall = recall, F1 = F1)
})
}


adjust_data <- function(data_m, metric_m, tools_nice_m, new_clus_id_m){
  d <- melt(data_m)
  colnames(d) <- c("value", "celltype", "score", "tool")
  d <- d[d$score == metric_m,]
  d$tool <- factor(d$tool, levels = rev(tools_nice_m[c(1,2,4,3,5)]))
  #plot_data$celltype <- factor(plot_data$celltype)
  #levels(plot_data$celltype)[c(5:7)] <- c("Platelets", "T-cells CD4+", "T-cells CD8+")
  d$celltype <- factor(d$celltype, levels = levels(new_clus_id_m$starsolo_celltypes))

  d[d$celltype!="unknown",]
}

get_plot <- function(data_m){
  ggplot(data_m, aes(x = celltype, y = tool, fill = value)) + 
    geom_tile() +
    geom_text(aes(label = signif(value, 3)), color = "grey95") +
    scale_y_discrete(expand = c(0.06, 0.01)) +
    scale_x_discrete(expand = c(0.06, 0.01)) +
    scale_fill_gradientn(colors = c("#deebf7", "#9ecae1", "#4292c6", "#08306b"), 
                       limits = c(0.0, 1.0)) +# ,mid = "#9ecae1", trans = "exp"
    #scale_fill_brewer(palette="Set1") +
    #guides(fill = guide_legend(title = "F1-score")) +
    theme_classic() +
    theme(text = element_text(size = 16),
          axis.line = element_blank(),
          axis.title = element_blank(),
          axis.text.x = element_text(angle = 60, vjust=+1.05, hjust = +1.1))

}


#### Create plots ####
p_bc_recall_prec_f1 <- list()
#### PBMC ####

new_clus_id<- list(cellranger6_celltypes = cluster2celltype(results_pbmc, "Cell Ranger 6"),
                   starsolo_celltypes = cluster2celltype(results_pbmc, "STARsolo"),
                   alevin_celltypes = cluster2celltype(results_pbmc, "Alevin"),
                   alevin_fry_celltypes = cluster2celltype(results_pbmc, "Alevin-fry"),
                   kallisto_celltypes = cluster2celltype(results_pbmc, "Kallisto"))


for (i in seq(integrated_data$PBMC)){
  integrated_data$PBMC[[i]]$celltypes <- integrated_data$PBMC[[i]]$seurat_clusters
  levels(integrated_data$PBMC[[i]]$celltypes) <- new_clus_id[[i]]
}

#### Plot data ####
categories <- get_category(integrated_data$PBMC, confident_cell_type_pbmc)
metrics_pbmc <- get_metrics(categories)
##### F1 #####
plot_data <- adjust_data(metrics_pbmc, "F1", tools_nice, new_clus_id)
p_bc_recall_prec_f1[["f1_pbmc"]] <- get_plot(plot_data) + 
  ggtitle("F1-score in PBMC") + theme(plot.title = element_text(hjust = .5))#, 
   #       subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA")
#p_bc_recall_prec_f1[["f1_pbmc"]]$labels$fill <-  "metric"
# ggsave("plots_revision/tmp/F1_score_PBMC.svg", p_bc_recall_prec_f1[["f1_pbmc"]], 
#        device = "svg", height = 8, width = 10)

##### precision #####
plot_data <- adjust_data(metrics_pbmc, "precision", tools_nice, new_clus_id)
p_bc_recall_prec_f1[["pbmc_prec"]] <- get_plot(plot_data) + 
  ggtitle("precision in PBMC") + theme(plot.title = element_text(hjust = .5))#, subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA")
#p_bc_recall_prec_f1[["pbmc_prec"]]$labels$fill <-  "precision"
#ggsave("plots_revision/tmp/precision_PBMC.svg", p_bc_recall_prec_f1[["pbmc_prec"]], 
#       device = "svg", height = 8, width = 10)

##### recall #####
plot_data <- adjust_data(metrics_pbmc, "recall", tools_nice, new_clus_id)
p_bc_recall_prec_f1[["pbmc_recall"]] <- get_plot(plot_data) + 
  ggtitle("recall in PBMC") + theme(plot.title = element_text(hjust = .5))#, subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA")
#p_bc_recall_prec_f1[["pbmc_recall"]]$labels$fill <-  "recall"
# ggsave("plots_revision/tmp/recall_PBMC.svg", p_bc_recall_prec_f1[["pbmc_recall"]], 
#        device = "svg", height = 8, width = 10)



#### Cardiac ####
new_clus_id <- list(cellranger6_celltypes = cluster2celltype(results_cardiac, "Cell Ranger 6"),
                   starsolo_celltypes = cluster2celltype(results_cardiac, "STARsolo"),
                   alevin_celltypes = cluster2celltype(results_cardiac, "Alevin"),
                   alevin_fry_celltypes = cluster2celltype(results_cardiac, "Alevin-fry"),
                   kallisto_celltypes = cluster2celltype(results_cardiac, "Kallisto"))


for (i in seq(integrated_data$Cardiac)){
  integrated_data$Cardiac[[i]]$celltypes <- integrated_data$Cardiac[[i]]$seurat_clusters
  levels(integrated_data$Cardiac[[i]]$celltypes) <- new_clus_id[[i]]
}


categories <- get_category(integrated_data$Cardiac, confident_cell_type_cardiac)
metrics_cardiac <- get_metrics(categories)

plot_data <- adjust_data(metrics_cardiac, "F1", tools_nice, new_clus_id)
p_bc_recall_prec_f1[["cardiac_f1"]] <- get_plot(plot_data) +
   ggtitle("F1-score in Cardiac") + theme(plot.title = element_text(hjust = .5))#, subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA")
#p_bc_recall_prec_f1[["cardiac_f1"]]$labels$fill <-  "F1-score"
# ggsave("plots_revision/tmp/F1_score_Cardiac.svg", p_bc_recall_prec_f1[["cardiac_f1"]], 
#        device = "svg", height = 8, width = 10)

##### precision #####
plot_data <- adjust_data(metrics_cardiac, "precision", tools_nice, new_clus_id)
p_bc_recall_prec_f1[["cardiac_prec"]] <- get_plot(plot_data) +
  ggtitle("Precision in Cardiac") + theme(plot.title = element_text(hjust = .5))#, subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA")
#p_bc_recall_prec_f1[["cardiac_prec"]]$labels$fill <-  "Precision"
# ggsave("plots_revision/tmp/precision_Cardiac.svg", p_bc_recall_prec_f1[["cardiac_prec"]], 
#        device = "svg", height = 8, width = 10)

##### recall #####
plot_data <- adjust_data(metrics_cardiac, "recall", tools_nice, new_clus_id)
p_bc_recall_prec_f1[["cardiac_recall"]] <- get_plot(plot_data) +
  ggtitle("Recall in Cardiac") + theme(plot.title = element_text(hjust = .5))#, subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA")
#p_bc_recall_prec_f1[["cardiac_recall"]]$labels$fill <-  "Recall"
# ggsave("plots_revision/tmp/recall_Cardiac.svg", p_bc_recall_prec_f1[["cardiac_recall"]], 
#        device = "svg", height = 8, width = 10)


ggpubr::ggarrange(plotlist = p_bc_recall_prec_f1[-c(4)], 
                                                 ncol = 2, nrow = 3,
                  labels = "AUTO", common.legend = T, legend = "right")

ggsave("plots_revision/Suppl_figure_f1_recall_prec.svg", height = 14, width = 12)




p_pbmc_prec <- p_pbmc_prec + theme(axis.text.x = element_blank()) + ylab("Precision") + ggtitle("PBMC")
p_cardiac_prec <- p_cardiac_prec + ggtitle("Cardiac") + theme(axis.text.y = element_blank(),
                                         axis.text.x = element_blank())
p_pbmc_recall <- p_pbmc_recall + ylab("Recall")
p_cardiac_recall <- p_cardiac_recall + theme(axis.text.y = element_blank())

p_prec_recall <- ggpubr::ggarrange(p_pbmc_prec, p_cardiac_prec, p_pbmc_recall, p_cardiac_recall,
                  labels = "AUTO", common.legend = T, legend = "right", align = "hv")
ggsave("plots_revision/precision_recall.svg", p_prec_recall, height = 9.5, width = 15)





#### Plot data ####
plot_data <- melt(metrics)
colnames(plot_data) <- c("value", "celltype", "score", "tool")
plot_data <- plot_data[plot_data$score == "F1",]
plot_data$tool <- factor(plot_data$tool, levels = rev(tools_nice))
plot_data$celltype <- factor(plot_data$celltype)
plot_data$celltype <- factor(plot_data$celltype, levels = levels(new_clus_id$cellranger6_celltypes)) 

plot_data <- plot_data[plot_data$celltype!="unknown",]
p_f1 <- ggplot(plot_data, aes(x = celltype, y = tool, fill = value)) + 
  geom_tile() +
  geom_text(aes(label = signif(value, 3)), color = "grey95") +
  scale_y_discrete(expand = c(0.06, 0.01)) +
  scale_x_discrete(expand = c(0.06, 0.01)) +
  scale_fill_gradientn(colors = c("#deebf7", "#9ecae1", "#4292c6", "#08306b"), 
                       limits = c(0.0, 1.0)) +# ,mid = "#9ecae1", trans = "exp"
  #scale_fill_brewer(palette="Set1") +
  #guides(fill = guide_legend(title = "F1-score")) +
  ggtitle("F1-score in Cardiac", subtitle = "Comparison between barcodes from Seurat clustering and\nbarcode consensus scheme from SCINA") +
  theme_classic() +
  theme(text = element_text(size = 16),
        axis.line = element_blank(),
        axis.title = element_blank(),
        axis.text.x = element_text(angle = 60, vjust=+1.05, hjust = +1.1))
  p_f1$labels$fill <- "F1-score"
p_f1

ggsave("plots_revision/tmp/F1_score_Cardiac.svg", p_f1, device = "svg", height = 8, width = 10)


```


```{r sankey and barplot for reviewer comment}
#### PBMC ####
# scina_pbmc <- scina_classi(integrated_data$PBMC, cell_type_genes_pbmc)
# 
# barcode_ident <- merge_list(scina_pbmc)
# confident_cell_type_pbmc_alluv <- major_vote(barcode_ident)
# 
# freq_bc_tool <- list()
# for (i in seq(ncol(barcode_ident)-1)){
#   name_tmp <- names(barcode_ident)[i+1]
#   bc_per_tool <- barcode_ident[(!is.na(barcode_ident[[name_tmp]])),"Row.names"]
#   freq_bc_tool[[name_tmp]] <- table(confident_cell_type_pbmc_alluv[row.names(confident_cell_type_pbmc_alluv) %in% bc_per_tool,"n_prim_type"])
# }
# plot_data <- melt(freq_bc_tool)
# plot_data$Var1 <- factor(plot_data$Var1, levels = c("5", "4","3", "2", "1"))
# plot_data$L1 <- factor(plot_data$L1, levels = c("Cellranger_cluster", "STARsolo_cluster", "Alevin_cluster", "Alevin-fry_cluster", "Kallisto_cluster"))
# levels(plot_data$L1) <- do.call(rbind, strsplit(levels(plot_data$L1), "_"))[,1]
# plot_data <- cbind(factor("all_barcodes"), plot_data)
# colnames(plot_data) <- c("all_barcodes", "n_bc", "n", "Tool")
# 
# 
# ggplot(plot_data, aes(axis1= all_barcodes, axis2 = n_bc, axis3 = Tool, y = n)) +
#   geom_alluvium(aes(fill = n_bc), show.legend = F) +
#   scale_fill_manual(values = c("#e41a1c", "#377eb8", "#4daf4a", "#984ea3", "#ff7f00")) +
#   geom_stratum() +
#   geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
#   theme_minimal() + 
#   ylab("Barcodes frequency") +
#   scale_x_discrete(limits = c("All barcodes", "Barcode occurance", "Tools")) +
#   theme(text = element_text(size = 14),
#         panel.grid.major.x = element_blank())
# 
# 
# ggsave("plots_revision/tmp/barcodes_per_tool_pbmc.svg")
# #### Cardiac ####
# 
# scina_cardiac <- scina_classi(integrated_data$Cardiac, cell_type_genes_cardiac)
# barcode_ident <- merge_list(scina_cardiac)
# confident_cell_type_cardiac_alluv <- major_vote(barcode_ident)
# 
# freq_bc_tool <- list()
# for (i in seq(ncol(barcode_ident)-1)){
#   name_tmp <- names(barcode_ident)[i+1]
#   bc_per_tool <- barcode_ident[(!is.na(barcode_ident[[name_tmp]])),"Row.names"]
#   freq_bc_tool[[name_tmp]] <- table(confident_cell_type_cardiac_alluv[row.names(confident_cell_type_cardiac_alluv) %in% bc_per_tool,"n_prim_type"])
# }
# plot_data <- melt(freq_bc_tool)
# plot_data$Var1 <- factor(plot_data$Var1, levels = c("5", "4","3", "2", "1"))
# plot_data$L1 <- factor(plot_data$L1, levels = c("Cellranger_cluster", "STARsolo_cluster", "Alevin_cluster", "Alevin-fry_cluster", "Kallisto_cluster"))
# levels(plot_data$L1) <- do.call(rbind, strsplit(levels(plot_data$L1), "_"))[,1]
# plot_data <- cbind(factor("all_barcodes"), plot_data)
# colnames(plot_data) <- c("all_barcodes", "n_bc", "n", "Tool")
# 
# ggplot(plot_data, aes(axis1= all_barcodes, axis2 = n_bc, axis3 = Tool, y = n)) +
#   geom_alluvium(aes(fill = n_bc), show.legend = F) +
#   scale_fill_manual(values = c("#e41a1c", "#377eb8", "#4daf4a", "#984ea3", "#ff7f00")) +
#   geom_stratum() +
#   geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
#   theme_minimal() + 
#   ylab("Barcodes frequency") +
#   scale_x_discrete(limits = c("All barcodes", "Barcode occurance", "Tools")) +
#   theme(text = element_text(size = 14),
#         panel.grid.major.x = element_blank())
# 
# ggsave("plots_revision/tmp/barcodes_per_tool_cardiac.svg")



#### Barplot####
# densed_bc <- lapply(bc_venn, function(x) {
#   bc_tmp <- confident_cell_type_cardiac_alluv[row.names(confident_cell_type_cardiac_alluv) %in% x,2]
#   table(bc_tmp)
# })

bc_venn <- lapply(integrated_data$PBMC, function(x) colnames(x))
bc_venn <- lapply(integrated_data$Cardiac, function(x) colnames(x))

bc_venn_melt <- melt(bc_venn)
bc_count <- table(bc_venn_melt$value)
bc_count <- melt(bc_count)

densed_bc2 <- lapply(bc_venn, function(x) {
  bc_tmp <- bc_count[bc_count$Var1 %in% x,2]
  table(bc_tmp)
})

plot_data <- melt(densed_bc2)
colnames(plot_data) <- c("n_tools", "value", "Tool")
plot_data$Tool <- factor(plot_data$Tool, levels = tools_nice)
plot_data$n_tools <- factor(plot_data$n_tools)

#alluvial_ts(plot_data[c(3,1,2)])

# ggplot(plot_data, aes(x = n_tools, y = value, fill = Tool)) + 
#   geom_area() +
#   scale_fill_brewer(palette = "Set2") +
#   theme_classic() +
#   theme(text = element_text(size = 14))

ggplot(plot_data, aes(x = Tool, y = value, fill = n_tools)) +
  geom_col() +
  scale_fill_brewer(palette = "Set2", direction = -1) +
  guides(fill = guide_legend(title = "Barcodes found\nby n tools")) +
  #scale_y_log10() +
  ylab("Frequency") +
  scale_y_continuous(expand = c(0, 0)) +
  theme_classic() +
  theme(text = element_text(size = 14),
        axis.title.x = element_blank())
  
  ggsave("plots_revision/tmp/revision_bc_per_tool_pbmc.svg")


# n_tmp <- plot_data[plot_data$bc_tmp==1,c(2,3)]
# colnames(n_tmp)[1] <- 1
# for (x in unique(plot_data$bc_tmp)[-c(1)]) {
#   n_tmp <- merge(n_tmp, plot_data[plot_data$bc_tmp==x,c(2,3),],by = "L1")
#   colnames(n_tmp)[x+1] <- x
# }


# ggplot(n_tmp, aes(axis1= 1, axis2 = 2, axis3 = 3, axis4 = 4, axis5 = 5, y = L1)) +
#   geom_alluvium() +
#   #scale_fill_manual(values = c("#e41a1c", "#377eb8", "#4daf4a", "#984ea3", "#ff7f00")) +
#   geom_stratum() +
#   geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
#   theme_minimal() + 
#   ylab("Barcodes frequency") +
#   scale_x_discrete(limits = c("All barcodes", "Barcode occurance", "Tools")) +
#   theme(text = element_text(size = 14),
#         panel.grid.major.x = element_blank())


####5er venn diagram ####

bc_venn <- lapply(integrated_data$PBMC, function(x) colnames(x))
p_venn5 <- ggVennDiagram(bc_venn)
ggsave("plots_revision/tmp/venn_5_pbmc.svg", p_venn5)
# upset
mat_upset <- make_comb_mat(bc_venn)
attr(mat_upset, "comb_size") <- log10(attr(mat_upset, "comb_size"))
upset_bc_revision <- UpSet(mat_upset, set_order = tools_nice,
      comb_col = c("#e41a1c",'#377eb8','#4daf4a','#984ea3','#ff7f00')[comb_degree(mat_upset)],
      comb_order = rev(order(comb_size(mat_upset))))

svg("plots_revision/tmp/upset_5_pbmc.svg", width = 6, height = 3)
  upset_bc_revision
dev.off()


bc_venn <- lapply(integrated_data$Cardiac, function(x) colnames(x))
p_venn5 <- ggVennDiagram(bc_venn) +
ggsave("plots_revision/tmp/venn_5_cardiac.svg", p_venn5)
# upset
mat_upset <- make_comb_mat(bc_venn)
attr(mat_upset, "comb_size") <- log10(attr(mat_upset, "comb_size"))
upset_bc_revision <- UpSet(mat_upset, set_order = tools_nice,
      comb_col = c("#e41a1c",'#377eb8','#4daf4a','#984ea3','#ff7f00')[comb_degree(mat_upset)])#,
      #comb_order = rev(order(comb_size(mat_upset))))

svg("plots_revision/tmp/upset_5_cardiac.svg", width = 6, height = 3)
  upset_bc_revision
dev.off()

```



```{r Suppl. figure 5 sankey plots}
suppl_figure5 <- list()
#### Set up functions ####
get_sankey_plot <- function(cell_type_genes_m, confident_cell_types_m, 
                            integrated_data_m, results_m, tools_nice_m=tools_nice){
  
  barcodes_scina <- get_scina_bc(cell_type_genes_m, confident_cell_types_m)
  barcodes_seurat <- get_seurat_bc(integrated_data_m)
  names(barcodes_seurat) <- tools_nice_m
  bc_to_celltype <- melt(barcodes_scina)
  
  p_tmp <- lapply(tools_nice_m, function(x) {
    bc_to_cluster <- melt(barcodes_seurat[[x]])
    cluster_to_celltype  <- merge(bc_to_cluster, bc_to_celltype, by = "value", all = T)
    colnames(cluster_to_celltype) <- c("value", "cluster", "celltype")
    cluster_to_celltype[is.na(cluster_to_celltype$cluster),"cluster"] <- "M.b."
    cluster_to_celltype[is.na(cluster_to_celltype$celltype),"celltype"] <- "unknown"
    cluster_per_celltype <- melt(table(cluster_to_celltype$cluster, cluster_to_celltype$celltype))
    cluster_per_celltype$Var2 <- factor(cluster_per_celltype$Var2, 
                                        levels = levels(results_m$SCINA_cell_types))
    cluster_per_celltype <- cluster_per_celltype[cluster_per_celltype$value>0,]
  #cluster_count <- cluster_per_celltype %>% group_by(Var1) %>% summarise(sum = sum(value))
  #celltype_count <- cluster_per_celltype %>% group_by(Var2) %>% summarise(sum = sum(value))
    
  ggplot(cluster_per_celltype, aes(y = value, axis1 = Var2, axis2 = Var1)) +
    geom_alluvium(aes(fill = Var2), width = 1/12, show.legend = F) +
    geom_stratum(width = 1/8, fill = "grey") +
    geom_text(stat = "stratum", aes(label = paste(stratum, after_stat(count), sep = "\n")), nudge_x = .1 ) +
    scale_fill_brewer(palette="Set1") +
    coord_cartesian(clip = 'off') +
    ggtitle(x) +
    theme_void() +
    theme(plot.title = element_text(hjust = .5), plot.margin = unit(c(1,12,2,12), "mm"))
  
})

ggpubr::ggarrange(plotlist = p_tmp, ncol = 3, nrow = 2)

}

#### Make plot for pbmc ####
suppl_figure5$pbmc <- get_sankey_plot(cell_type_genes_pbmc, 
                                      confident_cell_type_pbmc, 
                                      integrated_data$PBMC,
                                      results_pbmc)

#### Make plot for cardiac ####
suppl_figure5$Cardiac <- get_sankey_plot(cell_type_genes_cardiac, 
                                         confident_cell_type_cardiac, 
                                         integrated_data$Cardiac,
                                         results_cardiac)


#### Merge plots ####
ggpubr::ggarrange(plotlist = suppl_figure5, labels = "AUTO")
ggsave("plots_revision/Suppl_Figures/Suppl_figure_5_hf.svg", device = "svg", width = 18, height = 12)

```

```{r create DEGs}


figure4_plotlist <- list()
#### PBMC ####

# new_clus_id<- list(cellranger6_celltypes = cluster2celltype(results_pbmc, "Cell Ranger 6"),
#                    starsolo_celltypes = cluster2celltype(results_pbmc, "STARsolo"),
#                    alevin_celltypes = cluster2celltype(results_pbmc, "Alevin"),
#                    alevin_fry_celltypes = cluster2celltype(results_pbmc, "Alevin-fry"),
#                    kallisto_celltypes = cluster2celltype(results_pbmc, "Kallisto"))

new_clus_id<- list(
  cellranger6_celltypes=c("0"="T-cells CD4+","1"="T-cells CD4+","2"="Monocytes","3"="NK-cells","4"="NKT-cells",
                          "5"="B-cells","6"="B-cells","7"="T-cells CD8+","8"="Platelets"),
  
  starsolo_celltypes=c("0"="T-cells CD4+", "1"="Monocytes", "2"="T-cells CD4+", "3"="NK-cells", 
                       "4"="B-cells", "5"="B-cells", "6"="NKT-cells",
                       "7"="T-cells CD8+", "8"="unknown", "9"="Platelets", "10" = "unknown"),
  
  alevin_celltypes=c("0"="T-cells CD4+","1"="Monocytes","2"="T-cells CD4+","3"="NKT-cells","4"="B-cells",
                     "5"="B-cells","6"="T-cells CD8+","7"="NK-cells","8"="unknown","9"="unknown",
                     "10"="Platelets", "11"="unknown"),
  
  alevin_fry_celltypes = c("0"="T-cells CD4+", "1"="Monocytes", "2"="T-cells CD4+", "3"="NK-cells", 
                           "4"="NKT-cells","5"="B-cells","6"="B-cells", "7"="T-cells CD8+", 
                           "8"= "B-cells", "9"="Platelets", "10"="unknown"),
  
  kallisto_celltypes=c("0"="T-cells CD4+","1"="T-cells CD4+","2"="Monocytes","3"="NKT-cells","4"="NK-cells",
                       "5"="B-cells","6"="B-cells","7"="T-cells CD8+","8"="Platelets")
)


for (i in seq(integrated_data$PBMC)){
  integrated_data$PBMC[[i]]$celltypes <- integrated_data$PBMC[[i]]$seurat_clusters
  levels(integrated_data$PBMC[[i]]$celltypes) <- new_clus_id[[i]]
}

#remove unknown cell type
integrated_data$PBMC <- lapply(integrated_data$PBMC,function(x){
  Idents(x) <- "celltypes"
  x <- x[,!(x$celltypes %in% c("unknown"))] #"Endothelial cells",
  x
})

marker_genes_pbmc <- lapply(integrated_data$PBMC, function(x) {
  FindAllMarkers(x, assay = "SCT", only.pos = T, test.use = "bimod")
})

marker_genes_pbmc_sig <- lapply(marker_genes_pbmc, function(x) {
  x[x$p_val_adj<0.05,]
})

marker_genes_per_celltype <- lapply(marker_genes_pbmc_sig, function(tool) {
  names_tmp <- levels(tool$cluster)
  names(names_tmp) <- names_tmp
  lapply(names_tmp, function(celltype) {
    tool[tool$cluster==celltype, "gene"]
  })
})
  
#calcualte the jaccard index of DEGs between cellranger and the other tools for each celltype
overlap_degs <- list()
celltypes <- names(marker_genes_per_celltype$`Cell Ranger 6`)
for (i in celltypes){
  for (j in seq(marker_genes_per_celltype)){
    overlap_degs[[paste(i, names(marker_genes_per_celltype), sep = "_")[j] ]] <- jaccard(marker_genes_per_celltype[[1]][[i]],                                                                               marker_genes_per_celltype[[j]][[i]])
  }
}

overlap_degs_melt <- melt(overlap_degs)
overlap_degs_melt <- cbind(overlap_degs_melt, do.call(rbind, strsplit(overlap_degs_melt$L1, "_")))
colnames(overlap_degs_melt) <- c("overlap", "L1", "celltype","tool")
overlap_degs_melt$tool <- factor(overlap_degs_melt$tool, levels = rev(tools_nice[c(1,2,4,3,5)]))
overlap_degs_melt$celltype <- factor(overlap_degs_melt$celltype, levels = celltypes)
levels(overlap_degs_melt$celltype)[7] <- "Platelets"

figure4_plotlist[["deg_jac"]] <- ggplot(overlap_degs_melt, aes(x = celltype, y = tool)) +
  geom_tile(aes(fill = signif(overlap, 3))) +
  geom_text(aes(label = signif(overlap, 3)), color = "grey95") +
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_classic() +
  scale_fill_gradientn(colors = c("#deebf7", "#9ecae1", "#4292c6", "#08306b"),
                      ) +#limits = c(0.0, 1.0)
  ggtitle("Jaccard Index on DEGs. Cellranger vs. other tools") +
  xlab("SCINA cell types") +
  theme(text = element_text(size = 14), 
        axis.text.x = element_text(angle = 60, vjust = +1.05 , hjust = +1.1),
        axis.title = element_blank(),
        axis.line = element_blank(),
        plot.title = element_text(hjust = .5))

figure4_plotlist[["deg_jac"]]$labels$fill <- "Overlap"
#p_deg_overlap
#ggsave("plots_revision/Figure_4_heatmap_jac.svg", p_deg_overlap, height = 6, width = 8)
  
  
##### Heatmap of pearson correlation #####

logfc_per_celltype <- lapply(marker_genes_pbmc_sig, function(tool) {
  names_tmp <- levels(tool$cluster)
  names(names_tmp) <- names_tmp
  lapply(names_tmp, function(celltype) {
    tool[tool$cluster==celltype, "avg_log2FC"]
  })
})


cor_fc <- list()
celltypes <- names(logfc_per_celltype$`Cell Ranger 6`)
for (i in celltypes){
  for (j in seq(marker_genes_pbmc_sig)){
    x <- marker_genes_pbmc_sig[[1]]
    x <- x[x$cluster==i,c("avg_log2FC", "gene", "cluster")]
    y <- marker_genes_pbmc_sig[[j]]
    y <- y[y$cluster==i,c("avg_log2FC", "gene", "cluster")]
    xy <- merge(x,y,by = "gene", all = T)
    #print(str(xy))
    xy$avg_log2FC.x[is.na(xy$avg_log2FC.x)] <- 0
    xy$avg_log2FC.y[is.na(xy$avg_log2FC.y)] <- 0
    #print(str(xy))
    cor_fc[[paste(i, names(logfc_per_celltype), sep = "_")[j] ]] <- cor(xy$avg_log2FC.x, xy$avg_log2FC.y)
  }
}
#cor_fc[[paste(i, names(logfc_per_celltype), sep = "_")[j] ]] <- cor(logfc_per_celltype[[1]][[i]],                                                                               logfc_per_celltype[[j]][[i]])


cor_fc_melt <- melt(cor_fc)
cor_fc_melt <- cbind(cor_fc_melt, do.call(rbind, strsplit(cor_fc_melt$L1, "_")))
colnames(cor_fc_melt) <- c("overlap", "L1", "celltype","tool")
cor_fc_melt$tool <- factor(cor_fc_melt$tool, levels = rev(tools_nice[c(1,2,4,3,5)]))
cor_fc_melt$celltype <- factor(cor_fc_melt$celltype, levels = celltypes)
levels(cor_fc_melt$celltype)[7] <- "Platelets"

figure4_plotlist[["deg_cor"]] <- ggplot(cor_fc_melt, aes(x = celltype, y = tool)) +
  geom_tile(aes(fill = signif(overlap, 3))) +
  geom_text(aes(label = signif(overlap, 3)), color = "grey95") +
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_classic() +
  scale_fill_gradientn(colors = c("#deebf7", "#9ecae1", "#4292c6", "#08306b"),
                      ) +#limits = c(0.0, 1.0)
  ggtitle("Pearson correlation on mean log2FC. Cellranger vs. other tools") +
  xlab("SCINA cell types") +
  theme(text = element_text(size = 14), 
        axis.text.x = element_text(angle = 60, vjust = +1.05 , hjust = +1.1),
        axis.title = element_blank(),
        axis.line = element_blank(),
        plot.title = element_text(hjust = .5))

figure4_plotlist[["deg_cor"]]$labels$fill <- "Correlation (r)"
#p_deg_cor
#ggsave("plots_revision/Figure_4_heatmap_cor.svg", p_deg_cor, height = 6, width = 8)



############################
#### upset plot of DEGs ####


# remove degs above adjusted p-value >0.05
marker_genes_pbmc <- lapply(marker_genes_pbmc, function(x) {
  x <- x[x$p_val_adj < 0.05,7]
  x
})

names(marker_genes_pbmc) <- tools_nice
comb_mat <- make_comb_mat(marker_genes_pbmc, mode = "distinct")
attr(comb_mat, "comb_size") <- log10(attr(comb_mat, "comb_size")) # get a log10 scale y-axis

figure4_plotlist[["upset"]] <- UpSet(comb_mat, set_order = tools_nice,
      comb_col = c("#e41a1c",'#377eb8','#4daf4a','#984ea3','#ff7f00')[comb_degree(comb_mat)])

# svg("plots_revision/Figure_4_log10.svg", width = 6, height = 3)
#   UpSet(comb_mat,set_order = tools_nice,
#         comb_col = c("#e41a1c",'#377eb8','#4daf4a','#984ea3','#ff7f00')[comb_degree(comb_mat)])
# dev.off()



#################################
#### Scatter plots of log2FC ####
marker_genes_pbmc_cd4 <- lapply(integrated_data$PBMC, function(x) {
  d <- FindMarkers(x, ident.1 = "T-cell CD4+", assay = "SCT", only.pos = F, 
                   test.use = "bimod")
  d <- d[d$p_val_adj<0.01,]
  d
})

names(marker_genes_pbmc_cd4) <- gsub(" |-", "_", names(marker_genes_pbmc_cd4))

plot_list_FC <- list()
for (i in seq(marker_genes_pbmc_cd4)[-c(1)]){
  m_tmp <- merge(marker_genes_pbmc_cd4[[1]], marker_genes_pbmc_cd4[[i]], 
                 by = 0, all = T, suffixes = names(marker_genes_pbmc_cd4)[c(1,i)])
  m_tmp <- m_tmp[,grep("Row.names|p_val_adj|avg_log2FC",colnames(m_tmp))]

  name_tmp <- names(marker_genes_pbmc_cd4)[i]
  
  # assign significance for ggplot
  m_tmp$is_sign <- "Both"
  m_tmp[is.na(m_tmp$p_val_adjCell_Ranger_6),6] <- name_tmp
  m_tmp[is.na(m_tmp[paste0("p_val_adj", name_tmp )]),6] <- "Cellranger"
  
  # replace NA in logfc to zero
  m_tmp[is.na(m_tmp$avg_log2FCCell_Ranger_6),2] <- 0 
  m_tmp[is.na(m_tmp[paste0("avg_log2FC", name_tmp )]),4] <- 0
  #m_tmp[is.na(m_tmp$p_val_adjCell_Ranger_6),3] <- 1
  #m_tmp[is.na(m_tmp[paste0("avg_log2FC", name_tmp )]),3] <- 1
  m_tmp$is_sign <- factor(m_tmp$is_sign, levels = c("Both", "Cellranger", name_tmp))
  
  r_sqrt <- summary(lm(paste("avg_log2FCCell_Ranger_6", paste0("avg_log2FC", name_tmp), sep = " ~ "), data = m_tmp))$adj.r.squared


  plot_list_FC[[name_tmp]] <- ggplot(m_tmp, aes_string(x = "avg_log2FCCell_Ranger_6", 
                           y = paste0("avg_log2FC", name_tmp), 
                           color = "is_sign")) +
    geom_point() +
    annotate(geom = "text", label = paste0("adj. (R)^2 = ", round(r_sqrt,3)),
             x = -Inf, y = Inf, hjust = -0.1, vjust = 2) + #, vjust = "inward", hjust = "inward"
    ylab(paste0("Gene log2FC: ", name_tmp)) +
    xlab("Gene log2FC: Cellranger") +
    guides(color = guide_legend(title = "Significant in:")) +
    theme_article() +
    theme(text = element_text(size = 14),
          panel.grid.major = element_line(colour = "grey93"), 
          legend.position = "bottom")

}

# save heatmaps
p_fc <- ggpubr::ggarrange(plotlist = c(p_bc_recall_prec_f1["cardiac_f1"],
                                       figure4_plotlist[c("deg_jac", "deg_cor")]),
                          labels = "AUTO", align = "hv", ncol = 3, nrow = 1, legend = "bottom")
ggsave("plots_revision/tmp/DEG_log2FC_cellranger_vs_other2.svg", p_fc, 
       width = 18, height = 7)
  
svg("plots_revision/tmp/Upset_DEGs.svg", width = 6, height = 3)
  figure4_plotlist$upset
dev.off()

# save scatter plots
p_scatter_fc <- ggpubr::ggarrange(plotlist = plot_list_FC, 
                                  labels = c("D","E","F", "G"), 
                                  align = "hv", ncol = 2, nrow = 2)
ggsave("plots_revision/tmp/log2FC_cellranger_vs_other_scatter.svg",
       p_scatter_fc, height = 12,width = 12)
```


```{r Figure 4 DEG}

new_clus_id<- list(
  cellranger6_celltypes=c("0"="T-cells CD4+","1"="T-cells CD4+","2"="Monocytes","3"="NK-cells","4"="NKT-cells",
                          "5"="B-cells","6"="B-cells","7"="T-cell CD8+","8"="Endothelial cells"),
  
  starsolo_celltypes=c("0"="T-cells CD4+", "1"="Monocytes", "2"="T-cells CD4+", "3"="NK-cells", 
                       "4"="B-cells", "5"="B-cells", "6"="NKT-cells",
                       "7"="T-cell CD8+", "8"="unknown", "9"="Endothelial cells", "10" = "unknown"),
  
  alevin_celltypes=c("0"="T-cells CD4+","1"="Monocytes","2"="T-cells CD4+","3"="NKT-cells","4"="B-cells",
                     "5"="B-cells","6"="T-cell CD8+","7"="NK-cells","8"="unknown","9"="unknown",
                     "10"="Endothelial cells", "11"="unknown"),
  
  alevin_fry_celltypes = c("0"="T-cells CD4+", "1"="Monocytes", "2"="T-cells CD4+", "3"="NK-cells", 
                           "4"="NKT-cells","5"="B-cells","6"="B-cells", "7"="T-cell CD8+", 
                           "8"= "B-cells", "9"="Endothelial cells", "10"="unknown"),
  
  kallisto_celltypes=c("0"="T-cells CD4+","1"="T-cells CD4+","2"="Monocytes","3"="NKT-cells","4"="NK-cells",
                       "5"="B-cells","6"="B-cells","7"="T-cell CD8+","8"="Endothelial cells")
)

for (i in 1:5){
  integrated_data$PBMC[[i]]$celltypes <- integrated_data$PBMC[[i]]$seurat_clusters
  levels(integrated_data$PBMC[[i]]$celltypes) <- new_clus_id[[i]]
}

integrated_data$PBMC <- lapply(integrated_data$PBMC,function(x){
  Idents(x) <- "celltypes"
  x <- x[,!(x$celltypes %in% c("unknown"))] #"Endothelial cells",
  x
})


marker_genes_pbmc <- lapply(integrated_data$PBMC, function(x) {
  FindAllMarkers(x, assay = "SCT", only.pos = T, test.use = "bimod")
})


# remove degs above adjusted p-value >0.05
marker_genes_pbmc <- lapply(marker_genes_pbmc, function(x) {
  x <- x[x$p_val_adj < 0.05,7]
  x
})

names(marker_genes_pbmc) <- tools_nice
comb_mat <- make_comb_mat(marker_genes_pbmc, mode = "distinct")
attr(comb_mat, "comb_size") <- log10(attr(comb_mat, "comb_size")) # get a log10 scale y-axis
svg("plots_revision/Figure_4_log10.svg", width = 6, height = 3)
  UpSet(comb_mat,set_order = tools_nice,
        comb_col = c("#e41a1c",'#377eb8','#4daf4a','#984ea3','#ff7f00')[comb_degree(comb_mat)])
dev.off()


#### do heatmap of F1-score ####

marker_genes_pbmc_sig <- lapply(marker_genes_pbmc, function(x) {
  x[x$p_val_adj<0.05,]
})

marker_genes_per_celltype <- lapply(marker_genes_pbmc_sig, function(tool) {
  names_tmp <- levels(tool$cluster)
  names(names_tmp) <- names_tmp
  lapply(names_tmp, function(celltype) {
    tool[tool$cluster==celltype, "gene"]
  })
})
  
#calcualte the jaccard index of DEGs between cellranger and the other tools for each celltype
overlap_degs <- list()
celltypes <- names(marker_genes_per_celltype$`Cell Ranger 6`)
for (i in celltypes){
  for (j in seq(marker_genes_per_celltype)){
    overlap_degs[[paste(i, 
                        names(marker_genes_per_celltype), sep = "_")[j] ]] <- jaccard(marker_genes_per_celltype[[1]][[i]],                                                                               marker_genes_per_celltype[[j]][[i]])
  }
}

overlap_degs_melt <- melt(overlap_degs)
overlap_degs_melt <- cbind(overlap_degs_melt, do.call(rbind, strsplit(overlap_degs_melt$L1, "_")))
colnames(overlap_degs_melt) <- c("overlap", "L1", "celltype","tool")
overlap_degs_melt$tool <- factor(overlap_degs_melt$tool, levels = rev(c("Cell Ranger 6", "STARsolo", "Alevin-fry", "Alevin", "Kallisto")))
overlap_degs_melt$celltype <- factor(overlap_degs_melt$celltype, levels = c("T-cells CD4+", "Monocytes", "B-cells", "NK-cells", "NKT-cells", "T-cell CD8+", "Endothelial cells"))
levels(overlap_degs_melt$celltype)[7] <- "Platelets"

p_deg_overlap <- ggplot(overlap_degs_melt, aes(x = celltype, y = tool)) +
  geom_tile(aes(fill = signif(overlap, 3))) +
  geom_text(aes(label = signif(overlap, 3)), color = "grey95") +
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_classic() +
  scale_fill_gradient(low = "#bdc9e1", high =  "#045a8d") +
  ggtitle("Jaccard Index on DEGs. Cellranger vs. other tools") +
  xlab("SCINA cell types") +
  theme(text = element_text(size = 14), 
        axis.text.x = element_text(angle = 60, vjust = +1.05 , hjust = +1.1),
        axis.title = element_blank(),
        axis.line = element_blank(),
        plot.title = element_text(hjust = .5))

p_deg_overlap$labels$fill <- "Overlap"
p_deg_overlap
ggsave("plots_revision/Figure_4_heatmap_jac.svg", p_deg_overlap, height = 6, width = 8)
  
  
##### create heatplot of pearson correlation #####
marker_genes_pbmc_sig
  
logfc_per_celltype <- lapply(marker_genes_pbmc_sig, function(tool) {
  names_tmp <- levels(tool$cluster)
  names(names_tmp) <- names_tmp
  lapply(names_tmp, function(celltype) {
    tool[tool$cluster==celltype, "avg_log2FC"]
  })
})


cor_fc <- list()
celltypes <- names(logfc_per_celltype$`Cell Ranger 6`)
for (i in celltypes){
  for (j in seq(marker_genes_pbmc_sig)){
    x <- marker_genes_pbmc_sig[[1]]
    x <- x[x$cluster==i,c("avg_log2FC", "gene", "cluster")]
    y <- marker_genes_pbmc_sig[[j]]
    y <- y[y$cluster==i,c("avg_log2FC", "gene", "cluster")]
    xy <- merge(x,y,by = "gene", all = T)
    print(str(xy))
    xy$avg_log2FC.x[is.na(xy$avg_log2FC.x)] <- 0
    xy$avg_log2FC.y[is.na(xy$avg_log2FC.y)] <- 0
    print(str(xy))
    cor_fc[[paste(i, names(logfc_per_celltype), sep = "_")[j] ]] <- cor(xy$avg_log2FC.x, xy$avg_log2FC.y)
  }
}
cor_fc[[paste(i, names(logfc_per_celltype), sep = "_")[j] ]] <- cor(logfc_per_celltype[[1]][[i]],                                                                               logfc_per_celltype[[j]][[i]])


cor_fc_melt <- melt(cor_fc)
cor_fc_melt <- cbind(cor_fc_melt, do.call(rbind, strsplit(cor_fc_melt$L1, "_")))
colnames(cor_fc_melt) <- c("overlap", "L1", "celltype","tool")
cor_fc_melt$tool <- factor(cor_fc_melt$tool, levels = rev(c("Cell Ranger 6", "STARsolo", "Alevin-fry", "Alevin", "Kallisto")))
cor_fc_melt$celltype <- factor(cor_fc_melt$celltype, levels = c("T-cells CD4+", "Monocytes", "B-cells", "NK-cells", "NKT-cells", "T-cell CD8+", "Endothelial cells"))
levels(cor_fc_melt$celltype)[7] <- "Platelets"

p_deg_cor <- ggplot(cor_fc_melt, aes(x = celltype, y = tool)) +
  geom_tile(aes(fill = signif(overlap, 3))) +
  geom_text(aes(label = signif(overlap, 3)), color = "grey95") +
  scale_y_discrete(expand = c(0, 0)) +
  scale_x_discrete(expand = c(0, 0)) +
  theme_classic() +
  scale_fill_gradient(low = "#bdc9e1", high =  "#045a8d") +
  ggtitle("Pearson correlation on mean log2FC. Cellranger vs. other tools") +
  xlab("SCINA cell types") +
  theme(text = element_text(size = 14), 
        axis.text.x = element_text(angle = 60, vjust = +1.05 , hjust = +1.1),
        axis.title = element_blank(),
        axis.line = element_blank(),
        plot.title = element_text(hjust = .5))

p_deg_cor$labels$fill <- "Correlation (r)"
p_deg_cor
ggsave("plots_revision/Figure_4_heatmap_cor.svg", p_deg_cor, height = 6, width = 8)

p_deg_cor_overlap <- ggpubr::ggarrange(p_deg_overlap, p_deg_cor, labels = "AUTO", nrow = 1)
ggsave("plots_revision/Figure_4_heatmap_cor_overlap.svg",p_deg_cor_overlap, height = 6, width = 15)
```



```{r Suppl. Figure 8}
library(fishpond)


datasets <- c("pbmc_10x", "Endothelial", "Cardiac")
tools <- c("cellranger", "cellranger5", "starsolo", "alevin", "kallisto")
filtered <- c("results_filtered_anno", "results_unfiltered_anno")

filtered_data <- list()
filtered_data_seurat_obj <- list()
for (i in datasets){
  if (i == "Cardiac"){
    print(i)
    sample_names <- paste0(c("MF1701"), c(0,3:8))
  } else if(i == "Endothelial"){
    sample_names <- c("Brain", "Colon", "Heart", "Kidney", "Liver", "Lung", "muscle_EDL", "muscle_Soleus", "Small_Intestine", "Spleen", "Testis")
  } else if(i == "pbmc_10x"){
    sample_names <- paste0(c("sample"), c(1:4))
  }
  for (j in tools){
    print(j)
    if(j == "cellranger5" && i == "Endothelial") next
    for (k in filtered){
      #if (j =="cellranger5") k <- NULL
      for (l in sample_names){
        filtered_data[[paste(i,j,k,l,sep="__")]] <- read_data(i,j,k,l)
        filtered_data_seurat_obj[[paste(i,j,k,l,sep="__")]] <- CreateSeuratObject(filtered_data[[paste(i,j,k,l,sep="__")]],project = paste(i,j,k,l,sep="__"))
      }
    }
  }
}

filtered <- c("results_filtered_anno")
pbmc_cardiac_endo <- save_seurat_objects(datasets, tools, filtered)
names(pbmc_cardiac_endo) <- paste0(names(pbmc_cardiac_endo), "__filtered")
names(pbmc_cardiac_endo) <- gsub("pbmc_10x","PBMC", names(pbmc_cardiac_endo))

filtered <- c("results_unfiltered_anno")
pbmc_cardiac_endo_unfil <- save_seurat_objects(datasets, tools, filtered)
names(pbmc_cardiac_endo_unfil) <- paste0(names(pbmc_cardiac_endo_unfil), "__unfiltered")
names(pbmc_cardiac_endo_unfil) <- gsub("_unfiltered", "__unfiltered", names(pbmc_cardiac_endo_unfil))
names(pbmc_cardiac_endo_unfil) <- gsub("pbmc_10x","PBMC", names(pbmc_cardiac_endo_unfil))
pbmc_cardiac_endo2 <- c(pbmc_cardiac_endo, pbmc_cardiac_endo_unfil)

#### boxplots (A) for figure 8 ####

mt_content<- lapply(pbmc_cardiac_endo2, function(x) {
  x <- x[grep("mt-Rnr1|mt-Rnr2", row.names(x), invert = T, ignore.case = T) ,]
  x$percent.mt <- PercentageFeatureSet(x, pattern = "^MT-|^mt-")
  x$percent.mt
})

mt_long <- melt(mt_content)
mt_long <- cbind(mt_long, do.call(rbind, strsplit(mt_long$L1, "__")))
colnames(mt_long) <- c("value", "L1", "Sample", "Mapper", "Dataset", "Filtering")
mt_long$Dataset <- factor(mt_long$Dataset, levels = datasets_nice)
mt_long$Mapper <- factor(mt_long$Mapper, levels = tools)
levels(mt_long$Mapper) <- tools_nice
mt_long$Filtering <- factor(mt_long$Filtering)
levels(mt_long$Filtering) <- c("Filtered", "Unfiltered")


ggplot(mt_long, aes(x = Mapper, y = value, color = Filtering)) +
  geom_boxplot() +
  facet_grid(.~Dataset, scales = "free", space = "free") +
  ylab("Mitochondrial content per barcode (%)") +
  scale_y_log10() +
  theme_light() +
  scale_color_manual(values = c("#80b1d3", "#fdb462")) +
  theme(text = element_text(size = 15), 
        axis.text.x = element_text(angle = 60,vjust=+.85,hjust = +.9),
        panel.grid = element_blank())
  

ggsave("plots_revision/tmp/suppl_figure_8a_boxplot.png", device = "png", height = , width = )
ggsave("plots_revision/tmp/suppl_figure_8a_boxplot.svg", device = "svg", height = , width = )


#### barplot (B) for figure 8 #####
mt_content <- read.table("/media/Storage/additional_datasets/mt_content_read_counts.tsv")
colnames(mt_content) <- c("genes", "read_count", "Dataset", "read_fate")
mt_content$read_fate <- factor(mt_content$read_fate,levels = c("removed","kept"))
levels(mt_content$read_fate) <- c("yes", "no")
mt_content$Dataset <- factor(mt_content$Dataset, levels = c("pbmc","rosenthal"))
levels(mt_content$Dataset) <- c("PBMC", "Cardiac")

p_mt_change <- ggplot(mt_content,aes(x = genes, y = read_count, fill = read_fate)) +
  geom_col() +
  facet_grid(Dataset~., scales = "free", space = "free") +
  theme_light() +
  guides(fill = guide_legend(title = "Multi-\nmapped\nreads")) +
  theme(text = element_text(size = 19), axis.text.x = element_text(angle = 60,vjust=+.85,hjust = +.9),panel.grid = element_blank()) +
  scale_fill_manual(values = c("#fc8d62", "#8da0cb")) +
  scale_y_continuous(expand = c(0, 1000)) +
  annotation_logticks(sides = "l") +
  xlab("Mitochondrial genes") +
  ylab("Read count")

ggsave("plots_revision/tmp/figure8b_barplot_reads.svg", p_mt_change, device = "svg", height = 5,width = 10)

```


Cluster the cells to see if the cells from the mapping with the unfiltered 
annotation lie within the other clusters or form a specific new cluster 
indicating non valid cells.
```{r Suppl Figure 8c New barcodes}

# get new barcodes
get_mt_gene_count <- function(m_filtered_data){
  seurat_data <- CreateSeuratObject(m_filtered_data)
  seurat_data[["percent.mt"]] <- PercentageFeatureSet(seurat_data, pattern = "^mt-")
  seurat_data <- subset(seurat_data,subset=nFeature_RNA> 200 & nFeature_RNA < 2500 & percent.mt< 10)
  print(seurat_data)
  gene_names <- rownames(seurat_data)
  return(colnames(seurat_data))
  #gene_names_mt <- gene_names[grep("^mt-", gene_names)]
  #return(Matrix::rowSums(seurat_data[gene_names_mt]))
}

non_filtered_data_fig2 <- read_filtered_data("/media/ATLAS_NGS_storage/RalfSB/Results_Rosenthal/","Star_solo","MF17018","emptyDrops")
filtered_data_fig2 <- read_filtered_data("/media/Helios_scStorage/Ralf/MouseHom_AMI/Results_Rosenthal/","Star_solo","MF17018","emptyDrops")

mt_gene_count <- get_mt_gene_count(non_filtered_data_fig2)
mt_gene_count2 <- get_mt_gene_count(filtered_data_fig2)

# get the names of the new barcodes and the barcodes that are lost in the non-filtered dataset
new_barcodes <- mt_gene_count[!mt_gene_count %in% mt_gene_count2]
lost_barcodes <- mt_gene_count2[!mt_gene_count2 %in% mt_gene_count]
lost_barcodes_data <- filtered_data_fig2[,colnames(filtered_data_fig2) %in% lost_barcodes]
colnames(lost_barcodes_data) <- paste0(colnames(lost_barcodes_data),"lost_barcodes")

#change the barcode names in the two data sets, so that we can distinguish them later in the integration
barcode_names <- colnames(non_filtered_data_fig2)
barcode_names[barcode_names %in% new_barcodes] <- paste0(new_barcodes,"new_barcodes")
barcode_names[!barcode_names %in% paste0(new_barcodes,"new_barcodes")] <- paste0(barcode_names[!barcode_names %in% paste0(new_barcodes,"new_barcodes")],"non_filtered")
colnames(non_filtered_data_fig2) <- barcode_names

barcode_names <- colnames(filtered_data_fig2)
barcode_names<- paste0(barcode_names,"Filtered")
colnames(filtered_data_fig2) <- barcode_names
filtered_data_fig2 <- cbind(filtered_data_fig2,lost_barcodes_data)


seurat_data <- CreateSeuratObject(non_filtered_data_fig2)
mapping <- vector(mode = "character", length = ncol(seurat_data))
mapping[grep("non_filtered",colnames(seurat_data))] <- "Non-filtered"
mapping[grep("new_barcodes",colnames(seurat_data))] <- "New cells"

seurat_data[["mapping"]] <- mapping
seurat_data2 <- CreateSeuratObject(filtered_data_fig2)
mapping_filter <- vector(mode = "character", length = ncol(seurat_data2))
mapping_filter[grep("Filtered",colnames(seurat_data2))] <- "Filtered"
#mapping_filter[grep("lost_barcodes",colnames(seurat_data2))] <- "Lost cells"
seurat_data2[["mapping"]] <- mapping_filter
seurat_data[["percent.mt"]] <- PercentageFeatureSet(seurat_data, pattern = "^mt-")
seurat_data2[["percent.mt"]] <- PercentageFeatureSet(seurat_data2, pattern = "^mt-")


exp_mat_list <- list(non_filter_gtf = seurat_data, filtered_gtf = seurat_data2)
tic()
for (i in names(exp_mat_list)){
  exp_mat_list[[i]]<- SCTransform(exp_mat_list[[i]])
}
seurat_features <- SelectIntegrationFeatures(object.list = exp_mat_list,nfeatures = 2500)
exp_mat_list <- PrepSCTIntegration(object.list = exp_mat_list,anchor.features = seurat_features)
seurat_anchors <- FindIntegrationAnchors(object.list = exp_mat_list,normalization.method = "SCT",anchor.features = seurat_features)

seurat_integrated <- IntegrateData(anchorset = seurat_anchors, normalization.method = "SCT")
seurat_integrated <- RunPCA(seurat_integrated)
seurat_integrated <- RunUMAP(seurat_integrated, dims = 1:15)
toc()
#1079.615 sec elapsed ubuntu
seurat_integrated_filter <- subset(seurat_integrated,subset=mapping=="Filtered" & nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10)
#seurat_integrated_non_filter <- subset(seurat_integrated,subset=mapping=="Non-filtered" & nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 10)
seurat_integrated_lost_new <- subset(seurat_integrated, mapping=="New cells")

colnames_to_plot <- c(colnames(seurat_integrated_lost_new),colnames(seurat_integrated_filter))
subset_to_plot <- seurat_integrated[,colnames(seurat_integrated) %in% colnames_to_plot]

#subset_to_plot_filtered_only <- subset(subset_to_plot,subset=mapping=="Filtered")
#DimPlot(subset_to_plot_filtered_only,combine = F,pt.size = 1)

cluster_plot2 <- DimPlot(subset_to_plot,group.by = "mapping",combine = F,pt.size = 1,cols = c("#cbd5e8","#4daf4a","#e41a1c"),order =list("Lost cells","New cells","Filtered"))
cluster_plot2[[1]] <- cluster_plot2[[1]] + xlab("Dimension 1") + ylab("Dimension 2")
cluster_plot2[[1]] <- cluster_plot2[[1]] + theme(plot.margin = unit(x=c(10,5,0,4),units = "mm"))
levels(cluster_plot2[[1]]$data$mapping)[1:2] <- c("Cells from\nfiltered\nannotation","Additional\ncells from\nunfiltered\nannotation")


levels(cluster_plot[[1]]$data$mapping)[2] <- "New cells"
levels(cluster_plot[[1]]$data$mapping)[3] <- "Lost cells"
cluster_plot2[[1]] <- cluster_plot2[[1]] + ggtitle("")
cluster_plot[[1]][["theme"]][["text"]][["size"]] <- 27
cluster_plot[[1]][["theme"]][["legend.text"]][["size"]] <- 30
cluster_plot[[1]][["theme"]][["axis.text"]][["size"]] <- 20
#cluster_plot[[1]][["data"]][["mapping"]] <- factor(cluster_plot[[1]][["data"]][["mapping"]],levels = c("Filtered","New cells","Lost cells"))
png("plots_revision/tmp/suppl_figure8_umap_new_cells.png",width = 1300, height = 800)
cluster_plot
dev.off()

ggsave("plots_revsion/tmp/suppl_figure8_umap_new_cells.png", cluster_plot2[[1]], device="png", width = 8, height = 5)

```


```{r Suppl Figure 2}
datasets <- c("pbmc_10x", "Endothelial", "Cardiac")
tools <- c("cellranger", "cellranger5", "starsolo", "alevin", "kallisto")
filtered <- c("results_filtered_anno", "results_unfiltered_anno")

filtered_data <- list()
filtered_data_seurat_obj <- list()
for (i in datasets){
  if (i == "Cardiac"){
    print(i)
    sample_names <- paste0(c("MF1701"), c(0,3:8))
  } else if(i == "Endothelial"){
    sample_names <- c("Brain", "Colon", "Heart", "Kidney", "Liver", "Lung", "muscle_EDL", "muscle_Soleus", "Small_Intestine", "Spleen", "Testis")
  } else if(i == "pbmc_10x"){
    sample_names <- paste0(c("sample"), c(1:4))
  }
  for (j in tools){
    print(j)
    if(j == "cellranger5" && i == "Endothelial") next
    for (k in filtered){
      #if (j =="cellranger5") k <- NULL
      for (l in sample_names){
        filtered_data[[paste(i,j,k,l,sep="__")]] <- read_data(i,j,k,l,1)
        filtered_data_seurat_obj[[paste(i,j,k,l,sep="__")]] <- CreateSeuratObject(filtered_data[[paste(i,j,k,l,sep="__")]],project = paste(i,j,k,l,sep="__"))
        #filtered_data_seurat_obj[[paste(i,j,k,l,sep="__")]][["percent.mt"]] <- PercentageFeatureSet(filtered_data_seurat_obj[[paste(i,j,k,l,sep="__")]], pattern = "^MT-|^mt-")
      }
    }
  }
}

filtered <- c("results_filtered_anno", "results_unfiltered_anno")
pbmc_cardiac_endo <- save_seurat_objects(datasets, tools, filtered, 1)
names(pbmc_cardiac_endo) <- gsub("pbmc_10x","PBMC", names(pbmc_cardiac_endo))

all_genes_mouse <- read.table(paste0(github_path, "mapping/post_mapping/all_genes_mouse.tsv"), 
                              sep = "\t",colClasses = c(rep("character",4)))

all_genes_human <- read.delim(paste0(github_path, "mapping/post_mapping/all_genes_human.tsv"),
                              header = F,colClasses = c(rep("character",4)))

gene_counts <- lapply(filtered_data_seurat_obj, function(x) {
  gene_count <- rowSums(x)
  gene_count <- gene_count[gene_count > 0]
  gene_count
})

pbmc_biotypes <- lapply(gene_counts[grep("pbmc", names(gene_counts))], function(x) {
  if (grepl("\\.",names((x))[1]) ){
    biotype <- all_genes_human[all_genes_human$V1 %in% names(x),]
  }else {
    biotype <- all_genes_human[all_genes_human$V2 %in% names(x),]
  }
  biotype$V4
})

cardiac_endo_biotypes <- lapply(gene_counts[grep("Cardiac|Endothelial", names(gene_counts))], function(x) {
  if (grepl("\\.",names((x))[1]) ){
    biotype <- all_genes_mouse[all_genes_mouse$V1 %in% names(x),]
  }else {
    biotype <- all_genes_mouse[all_genes_mouse$V2 %in% names(x),]
  }
  biotype$V4
})

biotypes <- c(pbmc_biotypes, cardiac_endo_biotypes)

biotypes_long <- melt(biotypes)

biotypes_long <- cbind(biotypes_long, do.call(rbind, strsplit(biotypes_long$L1,"__")))
colnames(biotypes_long) <- c("value", "L1", "Dataset", "Mapper", "Filtering", "Sample")
biotypes_long$Dataset <- factor(biotypes_long$Dataset, levels = datasets)
levels(biotypes_long$Dataset) <- c("PBMC", "Endothelial", "Cardiac")
biotypes_long$Mapper <- factor(biotypes_long$Mapper, levels = tools)
levels(biotypes_long$Mapper) <- c("Cell Ranger", "Cell Ranger 5", "STARsolo", "Alevin", "Kallisto")
biotypes_long$Filtering <- factor(biotypes_long$Filtering)
levels(biotypes_long$Filtering) <- c("Filtered", "Unfiltered")
biotypes_long$value <- factor(biotypes_long$value)

# condense biotypes
levels(biotypes_long$value)[grep("TR_", levels(biotypes_long$value))] <- "TR-genes"
levels(biotypes_long$value)[grep("IG_", levels(biotypes_long$value))] <- "IG-genes"
levels(biotypes_long$value)[levels(biotypes_long$value) %in% c("polymorphic_pseudogene",
                                        "pseudogene",
                                        "unitary_pseudogene",
                                        "rRNA_pseudogene"
                                        )] <- "other_pseudogenes"
levels(biotypes_long$value)[levels(biotypes_long$value) %in% c("transcribed_unprocessed_pseudogene",
                                        "transcribed_processed_pseudogene",
                                        "transcribed_unitary_pseudogene")] <- "transcribed_pseudogenes"
levels(biotypes_long$value)[levels(biotypes_long$value) %in% c("translated_unprocessed_pseudogene",
                                        "translated_processed_pseudogene")] <- "translated_pseudogenes"
biotypes_long <- biotypes_long[!biotypes_long$value %in% c("Mt_tRNA","Mt_rRNA","scRNA","sRNA","ribozyme","scaRNA","vaultRNA","snoRNA","snRNA","rRNA"),] # remove unnecessary biotypes

biotypes_long$value <- droplevels(biotypes_long$value)
levels(biotypes_long$value) <- sub("_pseudogene|_pseudogenes", "\npseudogenes", levels(biotypes_long$value))
# change biotypes order
biotypes_long$value <- factor(biotypes_long$value, levels = c("protein_coding",
                                                                   "lncRNA",
                                                                   "IG_genes",
                                                                   "TR_genes",
                                                                   "TEC",
                                                                   "miRNA",
                                                                   "misc_RNA",
                                                                   "processed\npseudogenes",
                                                                   "unprocessed\npseudogenes",
                                                                   "transcribed\npseudogenes",
                                                                   "translated\npseudogenes",
                                                                   "other\npseudogenes"))
levels(biotypes_long$value)[c(1,3,4,7)] <- c("Protein coding", "IG genes", "TR genes", "misc RNA") 
biotypes_long_long <- melt(table(biotypes_long$value, biotypes_long$Mapper, biotypes_long$Dataset, biotypes_long$Filtering, biotypes_long$Sample))
colnames(biotypes_long_long) <- c("Biotype","Mapper", "Dataset",  "Filtering", "Sample","value")

biotypes_long_long_pbmc <- biotypes_long_long[biotypes_long_long$Dataset=="PBMC" & biotypes_long_long$Sample %in% paste0(c("sample"), c(1:4)),]
biotypes_long_long_cardiac <- biotypes_long_long[biotypes_long_long$Dataset=="Cardiac" & biotypes_long_long$Sample %in% paste0(c("MF1701"), c(0,3:8)),]
biotypes_long_long_endo <- biotypes_long_long[biotypes_long_long$Dataset=="Endothelial" & biotypes_long_long$Sample %in% c("Brain", "Colon", "Heart", "Kidney", "Liver", "Lung", "muscle_EDL", "muscle_Soleus", "Small_Intestine", "Spleen", "Testis"),]
biotypes_long_long_endo <- biotypes_long_long_endo[biotypes_long_long_endo$Mapper %in% c("Cell Ranger", "STARsolo", "Alevin", "Kallisto"),]

biotypes_long_long2 <- rbind(biotypes_long_long_pbmc, biotypes_long_long_cardiac, biotypes_long_long_endo)
#biotypes_long_long <- biotypes_long_long[biotypes_long_long$value>0,]


biotypes_long_long_aggr <- aggregate(value ~ Mapper + Dataset + Filtering + Biotype, data = biotypes_long_long2, mean)


ggplot(biotypes_long_long_aggr, aes(x = Biotype, y = value, color = Mapper, shape = Filtering)) +
  geom_point(position=position_dodge(width = 1), size = 3, alpha = 0.8) +
  geom_vline(xintercept = seq(levels(biotypes_long_long_aggr$Biotype))+.5, color = "grey90") +
  facet_grid(Dataset~.) +
  scale_shape_manual(values=c(19, 14)) +
  scale_color_brewer(palette = "Set2") +
  theme_light() +
  guides(shape = guide_legend(title = "Annotation")) +
  theme(text = element_text(size = 17),
        axis.text.x = element_text(angle = 60, vjust = +1, hjust = +1.1),
        panel.grid = element_blank())


ggsave("plots_revision/second_submission_plots/Suppl_Figures/Suppl_figure_3/Suppl_figure_3_CR5.svg", device = "svg", width = 13, height = 5)

```

```{r Suppl. Figure 4 Heatmaps}
# Graph for Vmn gene family will not be created because in humans it is
# not highly expressed. You do not see much in the heatmap.

p_heat <- lapply(pbmc_integrated$PBMC, function(x) {
  DefaultAssay(x) <-"RNA"
  olfr_genes <- Matrix::rowSums(x[grep("^OR[0-9]{1}[A-Z]{1}[0-9]+", row.names(x))])
  olfr_genes <- olfr_genes[olfr_genes>0]
  DoHeatmap(x, assay = "RNA", slot = "data", features = names(olfr_genes)) +
    scale_fill_gradientn(colors = c("#377eb8","#ff7f00", "#ff7f00", "#ff7f00")) + 
    scale_y_discrete(guide = guide_axis(check.overlap = TRUE))
})

p_heat_arranged <- ggpubr::ggarrange(plotlist = p_heat, labels = "AUTO", common.legend = T, legend = "right")

ggsave("plots_revision/Suppl_figure_42.png", p_heat_arranged, device = "png", height = 10, width = 16)


```




